{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This is a draft"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import requests\n",
    "import json\n",
    "import pandas as pd\n",
    "import myvariant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input\n",
    "#user_input = input('Please input file PATH. ')\n",
    "CSV_input = '/Users/manolodominguez/Desktop/git-repos/STP_mini_projects/Igor-Manuel/Igor-Manuel/csv_file.csv'\n",
    "\n",
    "# input to pd data frame\n",
    "df = pd.read_csv(CSV_input) \n",
    "\n",
    "# An extra copy will be neccesary for manipulation.\n",
    "df_draft = pd.read_csv(CSV_input) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Here we read and found what we want from the CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{11: 'rs268', 56: 'rs56116432', 57: 'rs334'}\n"
     ]
    }
   ],
   "source": [
    "##### Reading CSV file values and looking for variants IDs ######\n",
    "\n",
    "# Find Variant ID (rs000000) in CSV\n",
    "# \\d+ is neccesary in case the line find a rs+something. rs\\d+ looks for rs+ numbers\n",
    "rs = df_draft[df_draft.apply(lambda x:x.str.contains(\"rs\\d+\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "\n",
    "# Now, we save the results found in a dict key=index and value=variand ID\n",
    "if rs.empty == False:\n",
    "    ind = rs.index.to_list()\n",
    "    vals = list(rs.stack().values)\n",
    "    row2rs = dict(zip(ind, vals))\n",
    "    print(row2rs)\n",
    "    # We need to remove the row where rs has been found to avoid repetitions\n",
    "    # In case in same row more than one kind of ID Variant is stored (i.e rs# and NM_#)   \n",
    "    for index, rs  in row2rs.items(): \n",
    "        # This will be done in df_draft\n",
    "        df_draft = df_draft.drop(index)\n",
    "    \n",
    "## Same thing with other ID variants\n",
    "\n",
    "# Here with Variant ID (NM_0000000) in CSV\n",
    "NM = df_draft[df_draft.apply(lambda x:x.str.contains(\"NM_\\d+\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "\n",
    "if NM.empty == False:\n",
    "    ind = NM.index.to_list()\n",
    "    vals = list(NM.stack().values)\n",
    "    row2NM = dict(zip(ind, vals))\n",
    "    print(row2NM)\n",
    "    \n",
    "    for index, NM  in row2NM.items(): \n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "# Here with Variant ID (NP_0000000) in CSV\n",
    "NP = df_draft[df_draft.apply(lambda x:x.str.contains(\"NP_\\d+\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "\n",
    "if NP.empty == False:\n",
    "    ind = NP.index.to_list()\n",
    "    vals = list(NP.stack().values)\n",
    "    row2NP = dict(zip(ind, vals))\n",
    "    print(row2NP)\n",
    "    \n",
    "    for index, NP  in row2NP.items(): \n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "# Here with ClinVar field (RCV#) in CSV\n",
    "RCV = df_draft[df_draft.apply(lambda x:x.str.contains(\"RCV\\d+\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "    \n",
    "if RCV.empty == False:\n",
    "    ind = RCV.index.to_list()\n",
    "    vals = list(RCV.stack().values)\n",
    "    row2RCV = dict(zip(ind, vals))\n",
    "    print(row2RCV)\n",
    "    \n",
    "    for index, NP  in row2NP.items(): \n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "\n",
    "#https://rest.ensembl.org/documentation/info/vep_hgvs_get\n",
    "# If you see the output of this input you will agree with me that this is one common problem "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{11: 'rs268', 56: 'rs56116432', 57: 'rs334'}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2rs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Here we use API for requesting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensemble REST API\n",
    "server = \"http://rest.ensembl.org\"\n",
    "\n",
    "#Empty dict\n",
    "rs2features ={}\n",
    "\n",
    "for index, rs  in row2rs.items(): \n",
    "    \n",
    "    # Rows where substring 'rs' has been found need to be delete to avoid repetition\n",
    "    # This will be done in df_draft\n",
    "    df_draft = df_draft.drop(index)\n",
    "    \n",
    "    # Request each rs found\n",
    "    ext = \"/variation/human/{}?\".format(rs)\n",
    "    r = requests.get(server+ext, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    decoded = r.json()\n",
    "\n",
    "    # Dict of dict\n",
    "    rs2features[rs] = decoded\n",
    "# We need to filter this data. This data can be used to fill the ultimate VCF\n",
    "# Or can be used also to verify that the values of the CSV are ook\n",
    "# For example gene name or coodinates\n",
    "    \n",
    "# This step made here with variant ID must be made with ... "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rs268': {'MAF': 0.005192,\n",
       "  'most_severe_consequence': 'missense_variant',\n",
       "  'clinical_significance': ['uncertain significance',\n",
       "   'pathogenic',\n",
       "   'risk factor'],\n",
       "  'synonyms': ['VCV000001550',\n",
       "   'RCV000001615',\n",
       "   'RCV000988041',\n",
       "   'RCV000781944',\n",
       "   'VAR_004239',\n",
       "   '609708.0033',\n",
       "   'NM_000237.3:c.953A>G',\n",
       "   'NP_000228.1:p.Asn318Ser',\n",
       "   'NM_000237.2:c.953A>G',\n",
       "   'rs52818902',\n",
       "   'rs386571803',\n",
       "   'rs17850737'],\n",
       "  'evidence': ['Frequency',\n",
       "   '1000Genomes',\n",
       "   'Cited',\n",
       "   'ESP',\n",
       "   'Phenotype_or_Disease',\n",
       "   'ExAC',\n",
       "   'TOPMed',\n",
       "   'gnomAD'],\n",
       "  'source': 'Variants (including SNPs and indels) imported from dbSNP',\n",
       "  'mappings': [{'assembly_name': 'GRCh38',\n",
       "    'strand': 1,\n",
       "    'start': 19956018,\n",
       "    'end': 19956018,\n",
       "    'seq_region_name': '8',\n",
       "    'location': '8:19956018-19956018',\n",
       "    'coord_system': 'chromosome',\n",
       "    'allele_string': 'A/G',\n",
       "    'ancestral_allele': 'A'}],\n",
       "  'name': 'rs268',\n",
       "  'ambiguity': 'R',\n",
       "  'minor_allele': 'G',\n",
       "  'var_class': 'SNP'},\n",
       " 'rs56116432': {'mappings': [{'coord_system': 'chromosome',\n",
       "    'assembly_name': 'GRCh38',\n",
       "    'seq_region_name': '9',\n",
       "    'location': '9:133256042-133256042',\n",
       "    'start': 133256042,\n",
       "    'strand': 1,\n",
       "    'end': 133256042,\n",
       "    'ancestral_allele': 'C',\n",
       "    'allele_string': 'C/T'},\n",
       "   {'allele_string': 'C/T',\n",
       "    'end': 133256189,\n",
       "    'ancestral_allele': None,\n",
       "    'start': 133256189,\n",
       "    'strand': 1,\n",
       "    'location': 'CHR_HG2030_PATCH:133256189-133256189',\n",
       "    'seq_region_name': 'CHR_HG2030_PATCH',\n",
       "    'assembly_name': 'GRCh38',\n",
       "    'coord_system': 'chromosome'}],\n",
       "  'ambiguity': 'Y',\n",
       "  'synonyms': ['NP_065202.2:p.Gly230Asp',\n",
       "   'NM_020469.2:c.689G>A',\n",
       "   'NM_020469.3:c.689G>A'],\n",
       "  'MAF': 0.002596,\n",
       "  'name': 'rs56116432',\n",
       "  'evidence': ['Frequency', '1000Genomes', 'ESP', 'ExAC', 'TOPMed', 'gnomAD'],\n",
       "  'minor_allele': 'T',\n",
       "  'most_severe_consequence': 'missense_variant',\n",
       "  'var_class': 'SNP',\n",
       "  'source': 'Variants (including SNPs and indels) imported from dbSNP'},\n",
       " 'rs334': {'ambiguity': 'N',\n",
       "  'var_class': 'SNP',\n",
       "  'minor_allele': 'A',\n",
       "  'name': 'rs334',\n",
       "  'mappings': [{'allele_string': 'T/A/C/G',\n",
       "    'ancestral_allele': 'T',\n",
       "    'location': '11:5227002-5227002',\n",
       "    'coord_system': 'chromosome',\n",
       "    'start': 5227002,\n",
       "    'end': 5227002,\n",
       "    'seq_region_name': '11',\n",
       "    'assembly_name': 'GRCh38',\n",
       "    'strand': 1}],\n",
       "  'evidence': ['Frequency',\n",
       "   '1000Genomes',\n",
       "   'Cited',\n",
       "   'ESP',\n",
       "   'Phenotype_or_Disease',\n",
       "   'ExAC',\n",
       "   'TOPMed',\n",
       "   'gnomAD'],\n",
       "  'source': 'Variants (including SNPs and indels) imported from dbSNP',\n",
       "  'synonyms': ['NM_000518.5:c.20A>G',\n",
       "   'NM_000518.4:c.20A>T',\n",
       "   'NP_000509.1:p.Glu7Val',\n",
       "   'NM_000518.5:c.20A>T',\n",
       "   'NM_000518.4:c.20A>G',\n",
       "   'NP_000509.1:p.Glu7Gly',\n",
       "   'NP_000509.1:p.Glu7Ala',\n",
       "   'NM_000518.4:c.20A>C',\n",
       "   'NM_000518.5:c.20A>C',\n",
       "   'rs34315575',\n",
       "   'rs77121243',\n",
       "   'rs3168321',\n",
       "   'VCV000446731',\n",
       "   'RCV000477892',\n",
       "   'RCV000016580',\n",
       "   'RCV000016573',\n",
       "   'RCV000016575',\n",
       "   'VCV000446735',\n",
       "   'VCV000446748',\n",
       "   'VCV000446730',\n",
       "   'RCV000016877',\n",
       "   'VCV000015333',\n",
       "   'VCV000446737',\n",
       "   'RCV000576548',\n",
       "   'RCV000224000',\n",
       "   'RCV000999922',\n",
       "   'RCV001104054',\n",
       "   'VCV000446736',\n",
       "   'RCV000030905',\n",
       "   'RCV000016286',\n",
       "   'RCV000016577',\n",
       "   'RCV000623118',\n",
       "   'RCV000016576',\n",
       "   'RCV000723337',\n",
       "   'VCV000446747',\n",
       "   'RCV000016579',\n",
       "   'RCV000016879',\n",
       "   'VCV000015175',\n",
       "   'VCV000446738',\n",
       "   'RCV000016352',\n",
       "   'RCV000016574',\n",
       "   '226',\n",
       "   '2580',\n",
       "   '228',\n",
       "   '141900.0085',\n",
       "   '141900.0243',\n",
       "   'VAR_002862'],\n",
       "  'clinical_significance': ['uncertain significance',\n",
       "   'likely benign',\n",
       "   'pathogenic',\n",
       "   'other',\n",
       "   'protective'],\n",
       "  'most_severe_consequence': 'missense_variant',\n",
       "  'MAF': 0.02736}}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rs2features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "mv = myvariant.MyVariantInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'_id': 'chr17:g.7578532A>G',\n",
       " '_version': 4,\n",
       " 'clinvar': {'_license': 'http://bit.ly/2SQdcI0',\n",
       "  'allele_id': 27396,\n",
       "  'alt': 'G',\n",
       "  'chrom': '17',\n",
       "  'cytogenic': '17p13.1',\n",
       "  'gene': {'id': '7157', 'symbol': 'TP53'},\n",
       "  'hg19': {'end': 7578532, 'start': 7578532},\n",
       "  'hg38': {'end': 7675214, 'start': 7675214},\n",
       "  'hgvs': {'coding': ['LRG_321t1:c.398T>C',\n",
       "    'LRG_321t2:c.398T>C',\n",
       "    'LRG_321t3:c.398T>C',\n",
       "    'LRG_321t4:c.398T>C',\n",
       "    'LRG_321t5:c.2T>C',\n",
       "    'LRG_321t6:c.2T>C',\n",
       "    'LRG_321t7:c.2T>C',\n",
       "    'LRG_321t8:c.281T>C',\n",
       "    'NM_000546.5:c.398T>C',\n",
       "    'NM_001126112.2:c.398T>C',\n",
       "    'NM_001126113.2:c.398T>C',\n",
       "    'NM_001126114.2:c.398T>C',\n",
       "    'NM_001126115.1:c.2T>C',\n",
       "    'NM_001126116.1:c.2T>C',\n",
       "    'NM_001126117.1:c.2T>C',\n",
       "    'NM_001126118.1:c.281T>C',\n",
       "    'NM_001276695.2:c.281T>C',\n",
       "    'NM_001276696.2:c.281T>C',\n",
       "    'NM_001276697.2:c.-80T>C',\n",
       "    'NM_001276698.2:c.-80T>C',\n",
       "    'NM_001276699.2:c.-80T>C',\n",
       "    'NM_001276760.2:c.281T>C',\n",
       "    'NM_001276761.2:c.281T>C'],\n",
       "   'genomic': ['LRG_321:g.17337T>C',\n",
       "    'NC_000017.10:g.7578532A>G',\n",
       "    'NC_000017.11:g.7675214A>G',\n",
       "    'NG_017013.2:g.17337T>C']},\n",
       "  'omim': '191170.0011',\n",
       "  'rcv': [{'accession': 'RCV000013151',\n",
       "    'clinical_significance': 'Pathogenic',\n",
       "    'conditions': {'identifiers': {'gene': '553989',\n",
       "      'medgen': 'C1835398',\n",
       "      'mondo': 'MONDO:0007903',\n",
       "      'omim': '151623',\n",
       "      'orphanet': '524'},\n",
       "     'name': 'Li-Fraumeni syndrome 1 (LFS)',\n",
       "     'synonyms': ['Li-Fraumeni syndrome 3']},\n",
       "    'last_evaluated': '1999-01-01',\n",
       "    'number_submitters': 1,\n",
       "    'origin': 'germline',\n",
       "    'preferred_name': 'NM_000546.5(TP53):c.398T>C (p.Met133Thr)',\n",
       "    'review_status': 'no assertion criteria provided'},\n",
       "   {'accession': 'RCV000492130',\n",
       "    'clinical_significance': 'Pathogenic',\n",
       "    'conditions': {'identifiers': {'medgen': 'C0027672',\n",
       "      'mondo': 'MONDO:0015356'},\n",
       "     'name': 'Hereditary cancer-predisposing syndrome',\n",
       "     'synonyms': ['Neoplastic Syndromes, Hereditary',\n",
       "      'Tumor predisposition',\n",
       "      'Cancer predisposition']},\n",
       "    'last_evaluated': '2016-11-28',\n",
       "    'number_submitters': 1,\n",
       "    'origin': 'germline',\n",
       "    'preferred_name': 'NM_000546.5(TP53):c.398T>C (p.Met133Thr)',\n",
       "    'review_status': 'criteria provided, single submitter'}],\n",
       "  'ref': 'A',\n",
       "  'rsid': 'rs28934873',\n",
       "  'type': 'single nucleotide variant',\n",
       "  'variant_id': 12357}}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mv.getvariant('chr17:g.7578532A>G', fields = 'clinvar')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
