{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CSV2VCF_v2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Requeritments:\n",
    "# 1) The first column must contain Gene_symbol\n",
    "# If your csv file doesnt have gene name or Gene_symbol\n",
    "# That is fine, in the report document, you will be inform that the gene_symbol\n",
    "# of your variant are wrong but the convertion will be perform anyway\n",
    "\n",
    "# 2) The second column must contain the coordinates of your variants like this (3:987654321)\n",
    "# Call this column hg19 if your are working with this building or\n",
    "# call this column hg38 if your are working with the hg38 building\n",
    "\n",
    "# 3) Third column must contain your variand ID\n",
    "# Variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's import all modules used in this script:\n",
    "# The next 19 lines imports the modules needed, if some module is not installed, a exception will installed the\n",
    "# package(s) and it will import the module(s).\n",
    "# This has been tested. When execute the script in new envs, this works\n",
    "modules = ['imp','sys', 'requests', 'chardet', 'json', 'pandas', 'numpy', 're','httplib2', 'liftover']\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "for library in modules:\n",
    "    try:\n",
    "        exec(\"import {module}\".format(module=library))\n",
    "    except:\n",
    "        subprocess.call(['pip', 'install', library])\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from liftover import get_lifter\n",
    "from liftover import ChainFile\n",
    "from datetime import date\n",
    "\n",
    "# For a reason I can't explain, requests module did not work sometimes when testing our script in a new envs\n",
    "# To avoid this, this is imported again.\n",
    "import requests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The ultimate goal of this programme is to be executable without any interection with the terminal\n",
    "# owing to, the motivation of this programme is to create a programme \n",
    "# to be used by users without strong bioinformatic background.\n",
    "# To achieve this, the script will be in a folder called CSP2VCF. \n",
    "# In this folder will contain the script and 2 folders, one called input and a second called output.\n",
    "# Idealy, the user only have to put the input file (a CSV file) in input folder\n",
    "# and get the output in the output folder.\n",
    "# The next lines allow this in any location of the user's computer\n",
    " \n",
    "# This read the file located in the folder input\n",
    "actual_path = os.getcwd()\n",
    "folder_input = '/input/'\n",
    "input_file_name = os.listdir(actual_path+folder_input)\n",
    "input_file_name= ''.join(input_file_name)\n",
    "CSV_input = actual_path+folder_input+input_file_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This is to avoid UnicodeDecodeError: 'utf-8'\n",
    "# If the excel document is saved as CSV, not utf-8 errors should occur\n",
    "# However, if the user doesn't do this, and some special characters are found in the input\n",
    "# This partially avoids the issue.\n",
    "with open(CSV_input, 'rb') as f:\n",
    "    result = chardet.detect(f.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>hg38</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "      <th>chro</th>\n",
       "      <th>location_hg38</th>\n",
       "      <th>chro_N</th>\n",
       "      <th>location_hg19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>chr3</td>\n",
       "      <td>100951675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>chr17</td>\n",
       "      <td>41219633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>chr15</td>\n",
       "      <td>48892422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>chr14</td>\n",
       "      <td>23896042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>chr9</td>\n",
       "      <td>136131429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TNNI1</td>\n",
       "      <td>1:201418478</td>\n",
       "      <td>chr1:g.201387606_201387614delAAAAAAAA</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>201418478</td>\n",
       "      <td>chr1</td>\n",
       "      <td>201387606</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344XX-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>8</td>\n",
       "      <td>19956018</td>\n",
       "      <td>chr8</td>\n",
       "      <td>19813529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>16</td>\n",
       "      <td>2088197</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2138198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene         hg38                             Transcript Quality  \\\n",
       "0   IMPG2  3:101232831                  NM_016247.3:c.3183C>G   34/64   \n",
       "1   BRCA1  17:43067616                  NM_007294.3:c.5066T>C  49/131   \n",
       "2    FBN1  15:48600225                   NM_000138.4:c.356G>A   46/99   \n",
       "3    MYH7  14:23426833                  NM_000257.3:c.1988G>A   8/102   \n",
       "4   BORIS  9:133256042                             rs56116432   1/200   \n",
       "5   TNNI1  1:201418478  chr1:g.201387606_201387614delAAAAAAAA  99/100   \n",
       "6   HLA-C   6:31271394        NM_002117.5:c.344-46_344-45insT   34/64   \n",
       "7   HLA-C   6:31271394      NM_002117.5:c.344XX-46_344-45insT   34/64   \n",
       "8   HLA-C   6:31271394                              rs9281300   34/64   \n",
       "9     LPL   8:19956018                   NM_000237.2:c.953A>G   34/64   \n",
       "10   TSC2   16:2088197                              rs1799758   34/64   \n",
       "\n",
       "     Filter chro  location_hg38 chro_N location_hg19  \n",
       "0      PASS    3      101232831   chr3     100951675  \n",
       "1      PASS   17       43067616  chr17      41219633  \n",
       "2      PASS   15       48600225  chr15      48892422  \n",
       "3   NO PASS   14       23426833  chr14      23896042  \n",
       "4   NO PASS    9      133256042   chr9     136131429  \n",
       "5      PASS    1      201418478   chr1     201387606  \n",
       "6      PASS    6       31271394   chr6      31239171  \n",
       "7      PASS    6       31271394   chr6      31239171  \n",
       "8      PASS    6       31271394   chr6      31239171  \n",
       "9      PASS    8       19956018   chr8      19813529  \n",
       "10     PASS   16        2088197  chr16       2138198  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now, we can read and convert the input in a panda df\n",
    "df_input = pd.read_csv(CSV_input, encoding=result['encoding'])\n",
    "\n",
    "# For future manipulation, we will need two copies of the input\n",
    "df_draft = df\n",
    "\n",
    "# Let's obtain the coordinates of  the varfiant in both buildings\n",
    "\n",
    "# Because we don't know which buildind the user will use\n",
    "# One requeritment will be to specifify the name of the second column \n",
    "# as hg38 or hg19\n",
    "# so that we can generate the coordinates of both buidilng now\n",
    "\n",
    "if df.columns[1] == 'hg38':\n",
    "    # I  need to  convert the coordinates to hg19\n",
    "    # Let's get the coordinates from the input file  \n",
    "    df[\"chro\"] = df[\"hg38\"].apply(lambda x: x.split(\":\")[0])\n",
    "    df[\"location_hg38\"] = df[\"hg38\"].apply(lambda x: x.split(\":\")[1])\n",
    "    df['location_hg38'] = df['location_hg38'].apply(pd.to_numeric)\n",
    "    Chrom2Pos_hg19 = dict()\n",
    "    converter = get_lifter('hg38', 'hg19')\n",
    "    df['chro_N'] = ''\n",
    "    df['location_hg19'] = ''\n",
    "    for index, chro,location_hg38 in zip(df.index,df.chro, df.location_hg38): \n",
    "        df.loc[index, ['chro_N']] = [converter.query(chro, location_hg38)[0][0]]\n",
    "        df.loc[index, ['location_hg19']] = [converter.query(chro, location_hg38)[0][1]]\n",
    "else:\n",
    "    #The opposite\n",
    "    df[\"chro\"] = df[\"hg19\"].apply(lambda x: x.split(\":\")[0])\n",
    "    df[\"location_hg19\"] = df[\"hg19\"].apply(lambda x: x.split(\":\")[1])\n",
    "    df['location_hg19'] = df['location_hg19'].apply(pd.to_numeric)\n",
    "    Chrom2Pos_hg19 = dict()\n",
    "    converter = get_lifter('hg19', 'hg38')\n",
    "    df['chro_N'] = ''\n",
    "    df['location_hg38'] = ''\n",
    "    for index, chro,location_hg19 in zip(df.index,df.chro, df.location_hg19): \n",
    "        df.loc[index, ['chro_N']] = [converter.query(chro, location_hg19)[0][0]]\n",
    "        df.loc[index, ['location_hg38']] = [converter.query(chro, location_hg19)[0][1]]\n",
    "\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>hg38</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "      <th>chro</th>\n",
       "      <th>location_hg38</th>\n",
       "      <th>chro_N</th>\n",
       "      <th>location_hg19</th>\n",
       "      <th>ref:c.</th>\n",
       "      <th>alt:c.</th>\n",
       "      <th>ref:g.</th>\n",
       "      <th>alt:g.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>chr3</td>\n",
       "      <td>100951675</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>chr17</td>\n",
       "      <td>41219633</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>chr15</td>\n",
       "      <td>48892422</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>chr14</td>\n",
       "      <td>23896042</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>chr9</td>\n",
       "      <td>136131429</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TNNI1</td>\n",
       "      <td>1:201418478</td>\n",
       "      <td>chr1:g.201387606_201387614delAAAAAAAA</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>201418478</td>\n",
       "      <td>chr1</td>\n",
       "      <td>201387606</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344XX-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>8</td>\n",
       "      <td>19956018</td>\n",
       "      <td>chr8</td>\n",
       "      <td>19813529</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>16</td>\n",
       "      <td>2088197</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2138198</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene         hg38                             Transcript Quality  \\\n",
       "0   IMPG2  3:101232831                  NM_016247.3:c.3183C>G   34/64   \n",
       "1   BRCA1  17:43067616                  NM_007294.3:c.5066T>C  49/131   \n",
       "2    FBN1  15:48600225                   NM_000138.4:c.356G>A   46/99   \n",
       "3    MYH7  14:23426833                  NM_000257.3:c.1988G>A   8/102   \n",
       "4   BORIS  9:133256042                             rs56116432   1/200   \n",
       "5   TNNI1  1:201418478  chr1:g.201387606_201387614delAAAAAAAA  99/100   \n",
       "6   HLA-C   6:31271394        NM_002117.5:c.344-46_344-45insT   34/64   \n",
       "7   HLA-C   6:31271394      NM_002117.5:c.344XX-46_344-45insT   34/64   \n",
       "8   HLA-C   6:31271394                              rs9281300   34/64   \n",
       "9     LPL   8:19956018                   NM_000237.2:c.953A>G   34/64   \n",
       "10   TSC2   16:2088197                              rs1799758   34/64   \n",
       "\n",
       "     Filter chro  location_hg38 chro_N location_hg19 ref:c. alt:c. ref:g.  \\\n",
       "0      PASS    3      101232831   chr3     100951675      C      G      G   \n",
       "1      PASS   17       43067616  chr17      41219633      T      C      A   \n",
       "2      PASS   15       48600225  chr15      48892422      G      A      C   \n",
       "3   NO PASS   14       23426833  chr14      23896042      G      A      C   \n",
       "4   NO PASS    9      133256042   chr9     136131429                        \n",
       "5      PASS    1      201418478   chr1     201387606                        \n",
       "6      PASS    6       31271394   chr6      31239171                        \n",
       "7      PASS    6       31271394   chr6      31239171                        \n",
       "8      PASS    6       31271394   chr6      31239171                        \n",
       "9      PASS    8       19956018   chr8      19813529      A      G      T   \n",
       "10     PASS   16        2088197  chr16       2138198                        \n",
       "\n",
       "   alt:g.  \n",
       "0       C  \n",
       "1       G  \n",
       "2       T  \n",
       "3       T  \n",
       "4          \n",
       "5          \n",
       "6          \n",
       "7          \n",
       "8          \n",
       "9       C  \n",
       "10         "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I will need the ref and the alt to request the API\n",
    "df['ref:c.'] = ''\n",
    "df['alt:c.'] = ''\n",
    "for idx, value in df.iloc[:,2].iteritems():\n",
    "    if ':c.' in value and \">\" in value:\n",
    "        df.loc[idx, ['ref:c.']] = list(value.split(\">\")[0])[-1]\n",
    "        df.loc[idx, ['alt:c.']] = value.split(\">\")[1]\n",
    "        df['ref:g.'] = df['ref:c.'].replace({\"C\": \"G\", \"G\": \"C\", \"A\":\"T\", \"T\":\"A\"})\n",
    "        df['alt:g.'] = df['alt:c.'].replace({\"C\": \"G\", \"G\": \"C\", \"A\":\"T\", \"T\":\"A\"})\n",
    "        \n",
    "    if ':g.' in value and \">\" in value:\n",
    "        df.loc[idx, ['ref:g.']] = list(value.split(\">\")[0])[-1]\n",
    "        df.loc[idx, ['alt:g.']] = value.split(\">\")[1]\n",
    "        df['ref:c.'] = df['ref:g.'].replace({\"C\": \"G\", \"G\": \"C\", \"A\":\"T\", \"T\":\"A\"})\n",
    "        df['alt:c.'] = df['alt:g.'].replace({\"C\": \"G\", \"G\": \"C\", \"A\":\"T\", \"T\":\"A\"})\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>hg38</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "      <th>chro</th>\n",
       "      <th>location_hg38</th>\n",
       "      <th>chro_N</th>\n",
       "      <th>location_hg19</th>\n",
       "      <th>ref:c.</th>\n",
       "      <th>...</th>\n",
       "      <th>deletion</th>\n",
       "      <th>sum_ins</th>\n",
       "      <th>sum_del</th>\n",
       "      <th>location_ins</th>\n",
       "      <th>location_del</th>\n",
       "      <th>API_REF</th>\n",
       "      <th>API_ALT</th>\n",
       "      <th>API_GENE_SYM</th>\n",
       "      <th>API_LOCATION</th>\n",
       "      <th>insertion:g</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>chr3</td>\n",
       "      <td>100951675</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>100951675</td>\n",
       "      <td>100951675</td>\n",
       "      <td>100951675_100951675</td>\n",
       "      <td>100951675_100951675</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>chr17</td>\n",
       "      <td>41219633</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>41219633</td>\n",
       "      <td>41219633</td>\n",
       "      <td>41219633_41219633</td>\n",
       "      <td>41219633_41219633</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>chr15</td>\n",
       "      <td>48892422</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>48892422</td>\n",
       "      <td>48892422</td>\n",
       "      <td>48892422_48892422</td>\n",
       "      <td>48892422_48892422</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>chr14</td>\n",
       "      <td>23896042</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>23896042</td>\n",
       "      <td>23896042</td>\n",
       "      <td>23896042_23896042</td>\n",
       "      <td>23896042_23896042</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>chr9</td>\n",
       "      <td>136131429</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>136131429</td>\n",
       "      <td>136131429</td>\n",
       "      <td>136131429_136131429</td>\n",
       "      <td>136131429_136131429</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>TNNI1</td>\n",
       "      <td>1:201418478</td>\n",
       "      <td>chr1:g.201387606_201387614delAAAAAAAA</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1</td>\n",
       "      <td>201418478</td>\n",
       "      <td>chr1</td>\n",
       "      <td>201387606</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td>delAAAAAAAA</td>\n",
       "      <td>201387606</td>\n",
       "      <td>201387614</td>\n",
       "      <td>201387606_201387606</td>\n",
       "      <td>201387606_201387614</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>31239172</td>\n",
       "      <td>31239171</td>\n",
       "      <td>31239171_31239172</td>\n",
       "      <td>31239171_31239171</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>insT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344XX-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>31239172</td>\n",
       "      <td>31239171</td>\n",
       "      <td>31239171_31239172</td>\n",
       "      <td>31239171_31239171</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>insT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>31239171</td>\n",
       "      <td>31239171</td>\n",
       "      <td>31239171_31239171</td>\n",
       "      <td>31239171_31239171</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>8</td>\n",
       "      <td>19956018</td>\n",
       "      <td>chr8</td>\n",
       "      <td>19813529</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>19813529</td>\n",
       "      <td>19813529</td>\n",
       "      <td>19813529_19813529</td>\n",
       "      <td>19813529_19813529</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>16</td>\n",
       "      <td>2088197</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2138198</td>\n",
       "      <td></td>\n",
       "      <td>...</td>\n",
       "      <td></td>\n",
       "      <td>2138198</td>\n",
       "      <td>2138198</td>\n",
       "      <td>2138198_2138198</td>\n",
       "      <td>2138198_2138198</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene         hg38                             Transcript Quality  \\\n",
       "0   IMPG2  3:101232831                  NM_016247.3:c.3183C>G   34/64   \n",
       "1   BRCA1  17:43067616                  NM_007294.3:c.5066T>C  49/131   \n",
       "2    FBN1  15:48600225                   NM_000138.4:c.356G>A   46/99   \n",
       "3    MYH7  14:23426833                  NM_000257.3:c.1988G>A   8/102   \n",
       "4   BORIS  9:133256042                             rs56116432   1/200   \n",
       "5   TNNI1  1:201418478  chr1:g.201387606_201387614delAAAAAAAA  99/100   \n",
       "6   HLA-C   6:31271394        NM_002117.5:c.344-46_344-45insT   34/64   \n",
       "7   HLA-C   6:31271394      NM_002117.5:c.344XX-46_344-45insT   34/64   \n",
       "8   HLA-C   6:31271394                              rs9281300   34/64   \n",
       "9     LPL   8:19956018                   NM_000237.2:c.953A>G   34/64   \n",
       "10   TSC2   16:2088197                              rs1799758   34/64   \n",
       "\n",
       "     Filter chro  location_hg38 chro_N location_hg19 ref:c.  ...     deletion  \\\n",
       "0      PASS    3      101232831   chr3     100951675      C  ...                \n",
       "1      PASS   17       43067616  chr17      41219633      T  ...                \n",
       "2      PASS   15       48600225  chr15      48892422      G  ...                \n",
       "3   NO PASS   14       23426833  chr14      23896042      G  ...                \n",
       "4   NO PASS    9      133256042   chr9     136131429         ...                \n",
       "5      PASS    1      201418478   chr1     201387606         ...  delAAAAAAAA   \n",
       "6      PASS    6       31271394   chr6      31239171         ...                \n",
       "7      PASS    6       31271394   chr6      31239171         ...                \n",
       "8      PASS    6       31271394   chr6      31239171         ...                \n",
       "9      PASS    8       19956018   chr8      19813529      A  ...                \n",
       "10     PASS   16        2088197  chr16       2138198         ...                \n",
       "\n",
       "      sum_ins    sum_del         location_ins         location_del API_REF  \\\n",
       "0   100951675  100951675  100951675_100951675  100951675_100951675           \n",
       "1    41219633   41219633    41219633_41219633    41219633_41219633           \n",
       "2    48892422   48892422    48892422_48892422    48892422_48892422           \n",
       "3    23896042   23896042    23896042_23896042    23896042_23896042           \n",
       "4   136131429  136131429  136131429_136131429  136131429_136131429           \n",
       "5   201387606  201387614  201387606_201387606  201387606_201387614           \n",
       "6    31239172   31239171    31239171_31239172    31239171_31239171           \n",
       "7    31239172   31239171    31239171_31239172    31239171_31239171           \n",
       "8    31239171   31239171    31239171_31239171    31239171_31239171           \n",
       "9    19813529   19813529    19813529_19813529    19813529_19813529           \n",
       "10    2138198    2138198      2138198_2138198      2138198_2138198           \n",
       "\n",
       "   API_ALT API_GENE_SYM API_LOCATION insertion:g  \n",
       "0                                                 \n",
       "1                                                 \n",
       "2                                                 \n",
       "3                                                 \n",
       "4                                                 \n",
       "5                                                 \n",
       "6                                           insT  \n",
       "7                                           insT  \n",
       "8                                                 \n",
       "9                                                 \n",
       "10                                                \n",
       "\n",
       "[11 rows x 26 columns]"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# I will also need the location of del and ins\n",
    "\n",
    "df['ins'] = ''\n",
    "df['insertion:g'] = ''\n",
    "for idx, value in df.iloc[:,2].iteritems():\n",
    "    if 'ins' in value:\n",
    "        df.loc[idx, ['ins']] = len(list(value.split(\"ins\")[1]))\n",
    "        df.loc[idx, ['insertion:g']] = 'ins' + df.iloc[idx,[2]].apply(lambda x: x.split(\"ins\")[1])[0]\n",
    "        #df.loc[idx, ['insertion:g']] = df.loc[idx, ['insertion:g']].str.replace({\"C\": \"G\", \"G\": \"C\", \"A\":\"T\", \"T\":\"A\"})\n",
    "    else:\n",
    "        df.loc[idx, ['ins']] = 0\n",
    "df['del'] = ''\n",
    "df['deletion'] = ''\n",
    "for idx, value in df.iloc[:,2].iteritems():\n",
    "    if 'del' in value:\n",
    "        df.loc[idx, ['del']] = len(list(value.split(\"del\")[1]))\n",
    "        df.loc[idx, ['deletion']] = 'del' + df.iloc[idx,[2]].apply(lambda x: x.split(\"del\")[1])[0]\n",
    "    else:\n",
    "        df.loc[idx, ['del']] = 0\n",
    "        \n",
    "df['sum_ins'] = df['location_hg19'].astype(int) + df['ins'].astype(int)\n",
    "df['sum_del'] = df['location_hg19'].astype(int) + df['del'].astype(int)\n",
    "\n",
    "df['sum_ins'] = df['sum_ins'].astype(str)\n",
    "df['sum_del'] = df['sum_del'].astype(str)\n",
    "df['location_hg19'] = df['location_hg19'].astype(str)\n",
    "df['location_ins'] =  df[['location_hg19', 'sum_ins']].agg('_'.join, axis=1)\n",
    "df['location_del'] =  df[['location_hg19', 'sum_del']].agg('_'.join, axis=1)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I am going to create the columns where I am going to store the info I want from the API\n",
    "df['API_REF'] = ''\n",
    "df['API_ALT'] = ''\n",
    "df['API_GENE_SYM'] = ''\n",
    "df['API_LOCATION'] = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 413,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid character in identifier (<ipython-input-413-c53901f97331>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-413-c53901f97331>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    • Insertion example:chr2:g.17142_17143insA\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid character in identifier\n"
     ]
    }
   ],
   "source": [
    "• Insertion example:chr2:g.17142_17143insA\n",
    "Deletion example:chrMT:g.8271_8279del"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_draft = df_draft.applymap(str)\n",
    "rs = df_draft[df_draft.apply(lambda x:x.str.contains(\"rs\\d+\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "if rs.empty == False:\n",
    "    ind = rs.index.to_list()\n",
    "    vals = list(rs.stack().values)\n",
    "    row2rs = dict(zip(ind, vals))\n",
    "    # I have seen occasionally how users save more than one variant ID in the same row. \n",
    "    # To avoid repetition, I delete the row\n",
    "    for index, rs  in row2rs.items(): \n",
    "        # This will be done in df_draft\n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "\n",
    "# Second kind of variant ID is insertion\n",
    "insertion = df_draft[df_draft.apply(lambda x:x.str.contains(\"ins\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "if insertion.empty == False:\n",
    "    ind = insertion.index.to_list()\n",
    "    vals = list(insertion.stack().values)\n",
    "    row2insertion = dict(zip(ind, vals))\n",
    "    for index, insertion  in row2insertion.items(): \n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "\n",
    "# Third  kind of variant ID is deletion\n",
    "deletion = df_draft[df_draft.apply(lambda x:x.str.contains(\"del\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "if deletion.empty == False:\n",
    "    ind = deletion.index.to_list()\n",
    "    vals = list(deletion.stack().values)\n",
    "    row2deletion = dict(zip(ind, vals))\n",
    "    for index, deletion  in row2deletion.items(): \n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "\n",
    "subtitution = df_draft[df_draft.apply(lambda x:x.str.contains(\">\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "if subtitution.empty == False:\n",
    "    ind = subtitution.index.to_list()\n",
    "    vals = list(subtitution.stack().values)\n",
    "    row2subtitution = dict(zip(ind, vals))\n",
    "    for index, subtitution  in row2subtitution.items(): \n",
    "        df_draft = df_draft.drop(index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################\n",
    "##### Its time for requesting API#####\n",
    "######################################\n",
    "\n",
    "#### If the Variant ID is dbnsp RS123 ####\n",
    "\n",
    "row2gene_symbol = dict()\n",
    "row2CHROM = dict()\n",
    "row2POS = dict()\n",
    "row2ALT = dict()\n",
    "row2REF = dict()\n",
    "\n",
    "\n",
    "server = \"http://myvariant.info/v1/query?q=\"\n",
    "\n",
    "\n",
    "for row, rs in row2rs.items():\n",
    "    # Request API here\n",
    "    r = requests.get(server+rs)\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "    decoded = r.json()\n",
    "    # We select the info we want and store it in dict\n",
    "    \n",
    "    # For row2gene_symbol\n",
    "    row2gene_symbol[row]= decoded['hits'][0]['dbsnp']['gene']['symbol']\n",
    "\n",
    "    # For row2CHROM\n",
    "    row2CHROM[row] = decoded['hits'][0]['chrom']\n",
    "    \n",
    "    # For row2POS \n",
    "    row2POS[row] = decoded['hits'][0]['vcf']['position']\n",
    "    \n",
    "    # For row2ALT \n",
    "    row2ALT[row] = decoded['hits'][0]['vcf']['alt']\n",
    "    \n",
    "    # For row2REF\n",
    "    row2REF[row] = decoded['hits'][0]['vcf']['ref']\n",
    "        \n",
    "# Next idea, to provide the citations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### If the Variant ID is a SUBSTITUTION ####\n",
    "g = ':g.'\n",
    "R_S_A = '>'\n",
    "server = 'http://myvariant.info/v1/variant/'\n",
    "for idx, value in df.iloc[:,2].iteritems():\n",
    "    if ':c.' in value and \">\" in value:\n",
    "        API_input = server + df.loc[idx, ['chro_N']][0] + g + df.loc[idx, ['location_hg19']][0] + df.loc[idx, ['ref:g.']][0] + R_S_A + df.loc[idx, ['alt:g.']][0]\n",
    "        r = requests.get(API_input)\n",
    "        if r.status_code == 200:\n",
    "            decoded = r.json()\n",
    "            row2gene_symbol[idx]= decoded['cadd']['gene']['genename']\n",
    "            row2ALT[idx] = decoded['vcf']['alt']\n",
    "            row2REF[idx] = decoded['vcf']['ref']\n",
    "            row2POS[idx] = decoded['vcf']['position']\n",
    "        else:\n",
    "            row2gene_symbol[idx] = \"Variant not found in API\"\n",
    "            row2ALT[idx] = \"Variant not found in API\"\n",
    "            row2REF[idx] = \"Variant not found in API\"\n",
    "            row2POS[idx] = \"Variant not found in API\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### If the Variant ID is a DELETION ####\n",
    "\n",
    "for idx, value in df.iloc[:,2].iteritems():\n",
    "    if 'del' in value :\n",
    "        \n",
    "        API_input = server + df.loc[idx, ['chro_N']][0] + g + df.loc[idx, ['location_del']][0] + 'del'\n",
    "        r = requests.get(API_input)\n",
    "        if r.status_code == 200:\n",
    "            decoded = r.json()\n",
    "            row2gene_symbol[idx]= decoded['dbsnp']['gene']['symbol']\n",
    "            row2ALT[idx] = decoded['vcf']['alt']\n",
    "            row2REF[idx] = decoded['vcf']['ref']\n",
    "            row2POS[idx] = decoded['vcf']['position']\n",
    "        else:\n",
    "            print(API_input)\n",
    "            row2gene_symbol[idx] = \"Variant not found in API\"\n",
    "            row2ALT[idx] = \"Variant not found in API\"\n",
    "            row2REF[idx] = \"Variant not found in API\"\n",
    "            row2POS[idx] = \"Variant not found in API\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "http://myvariant.info/v1/variant/chr6:g.31239171_31239172ins\n",
      "http://myvariant.info/v1/variant/chr6:g.31239171_31239172ins\n"
     ]
    }
   ],
   "source": [
    "#### If the Variant ID is a INSERTION ####\n",
    "\n",
    "for idx, value in df.iloc[:,2].iteritems():\n",
    "    if 'ins' in value :\n",
    "        \n",
    "        API_input = server + df.loc[idx, ['chro_N']][0] + g + df.loc[idx, ['location_ins']][0] + 'ins'\n",
    "        r = requests.get(API_input)\n",
    "        if r.status_code == 200:\n",
    "            decoded = r.json()\n",
    "            row2gene_symbol[idx]= decoded['dbsnp']['gene']['symbol']\n",
    "            row2ALT[idx] = decoded['vcf']['alt']\n",
    "            row2REF[idx] = decoded['vcf']['ref']\n",
    "            row2POS[idx] = decoded['vcf']['position']\n",
    "        else:\n",
    "            print(API_input)\n",
    "            row2gene_symbol[idx] = \"Variant not found in API\"\n",
    "            row2ALT[idx] = \"Variant not found in API\"\n",
    "            row2REF[idx] = \"Variant not found in API\"\n",
    "            row2POS[idx] = \"Variant not found in API\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr16:g.2138198_2138198ins'"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[idx, ['chro_N']][0] + g + df.loc[idx, ['location_ins']][0] + 'ins'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'Index' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-75-ea942bf769f0>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m: 'Index' object is not callable"
     ]
    }
   ],
   "source": [
    "df.columns()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{4: '136131429', 9: '31239170', 11: '2138199'}"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chr2:g.17142_17143insA\n",
    "# If the variant ID is dbsnp (i.e. rs00001)\n",
    "# The only thing I have to do is to request this variant ID directly in My.variant API\n",
    "\n",
    "row2gene_symbol = dict()\n",
    "row2CHROM = dict()\n",
    "row2POS = dict()\n",
    "row2ALT = dict()\n",
    "row2REF = dict()\n",
    "\n",
    "\n",
    "server = \"http://myvariant.info/v1/query?q=\"\n",
    "\n",
    "\n",
    "for row, rs in row2rs.items():\n",
    "    # Request API here\n",
    "    r = requests.get(server+rs)\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "    decoded = r.json()\n",
    "    # We select the info we want and store it in dict\n",
    "    \n",
    "    # For row2gene_symbol\n",
    "    row2gene_symbol[row]= decoded['hits'][0]['dbsnp']['gene']['symbol']\n",
    "\n",
    "    # For row2CHROM\n",
    "    row2CHROM[row] = decoded['hits'][0]['chrom']\n",
    "    \n",
    "    # For row2POS \n",
    "    row2POS[row] = decoded['hits'][0]['vcf']['position']\n",
    "    \n",
    "    # For row2ALT \n",
    "    row2ALT[row] = decoded['hits'][0]['vcf']['alt']\n",
    "    \n",
    "    # For row2REF\n",
    "    row2REF[row] = decoded['hits'][0]['vcf']['ref']\n",
    "        \n",
    "# Next idea, to provide the citations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>hg38</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117199558</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344XX-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene         hg38                         Transcript Quality   Filter\n",
       "0   IMPG2  3:101232831              NM_016247.3:c.3183C>G   34/64     PASS\n",
       "1   BRCA1  17:43067616              NM_007294.3:c.5066T>C  49/131     PASS\n",
       "2    FBN1  15:48600225               NM_000138.4:c.356G>A   46/99     PASS\n",
       "3    MYH7  14:23426833              NM_000257.3:c.1988G>A   8/102  NO PASS\n",
       "4   BORIS  9:133256042                         rs56116432   1/200  NO PASS\n",
       "5    CFTR  7:117199558  ENST00000003084:c.1431_1433delTTC  99/100     PASS\n",
       "6    CFTR  7:117559502  ENST00000003084:c.1431_1433delTTC  99/100     PASS\n",
       "7   HLA-C   6:31271394    NM_002117.5:c.344-46_344-45insT   34/64     PASS\n",
       "8   HLA-C   6:31271394  NM_002117.5:c.344XX-46_344-45insT   34/64     PASS\n",
       "9   HLA-C   6:31271394                          rs9281300   34/64     PASS\n",
       "10    LPL   8:19956018               NM_000237.2:c.953A>G   34/64     PASS\n",
       "11   TSC2   16:2088197                          rs1799758   34/64     PASS"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Now, we do the same with other kind of variant IDs\n",
    "# With the dbSNP variant ID was very easy\n",
    "# however with other ID things become more complicated\n",
    "# I still want to used my.variant API\n",
    "# This API so far only works with hg19\n",
    "# That means that we need to convert the location if the user use hg38.\n",
    "# I want to be ask the user which genome building is going to use.\n",
    "# To know in which genomic construction the user is working, \n",
    "# we will ask  him/her (in requeritments) to change the name of the second column.\n",
    "# to \"hg19\" or \"hg38\"\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>hg38</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "      <th>chro</th>\n",
       "      <th>location_hg38</th>\n",
       "      <th>chro_N</th>\n",
       "      <th>location_hg_19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>chr3</td>\n",
       "      <td>100951675.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>chr17</td>\n",
       "      <td>41219633.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>chr15</td>\n",
       "      <td>48892422.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>chr14</td>\n",
       "      <td>23896042.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>chr9</td>\n",
       "      <td>136131429.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117199558</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>7</td>\n",
       "      <td>117199558</td>\n",
       "      <td>chr7</td>\n",
       "      <td>117199556.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>7</td>\n",
       "      <td>117559502</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr8</td>\n",
       "      <td>19813529.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344XX-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2138198.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>8</td>\n",
       "      <td>19956018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>16</td>\n",
       "      <td>2088197</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene         hg38                         Transcript Quality   Filter  \\\n",
       "0   IMPG2  3:101232831              NM_016247.3:c.3183C>G   34/64     PASS   \n",
       "1   BRCA1  17:43067616              NM_007294.3:c.5066T>C  49/131     PASS   \n",
       "2    FBN1  15:48600225               NM_000138.4:c.356G>A   46/99     PASS   \n",
       "3    MYH7  14:23426833              NM_000257.3:c.1988G>A   8/102  NO PASS   \n",
       "4   BORIS  9:133256042                         rs56116432   1/200  NO PASS   \n",
       "5    CFTR  7:117199558  ENST00000003084:c.1431_1433delTTC  99/100     PASS   \n",
       "6    CFTR  7:117559502  ENST00000003084:c.1431_1433delTTC  99/100     PASS   \n",
       "7   HLA-C   6:31271394    NM_002117.5:c.344-46_344-45insT   34/64     PASS   \n",
       "8   HLA-C   6:31271394  NM_002117.5:c.344XX-46_344-45insT   34/64     PASS   \n",
       "9   HLA-C   6:31271394                          rs9281300   34/64     PASS   \n",
       "10    LPL   8:19956018               NM_000237.2:c.953A>G   34/64     PASS   \n",
       "11   TSC2   16:2088197                          rs1799758   34/64     PASS   \n",
       "\n",
       "   chro  location_hg38 chro_N  location_hg_19  \n",
       "0     3      101232831   chr3     100951675.0  \n",
       "1    17       43067616  chr17      41219633.0  \n",
       "2    15       48600225  chr15      48892422.0  \n",
       "3    14       23426833  chr14      23896042.0  \n",
       "4     9      133256042   chr9     136131429.0  \n",
       "5     7      117199558   chr7     117199556.0  \n",
       "6     7      117559502   chr6      31239171.0  \n",
       "7     6       31271394   chr8      19813529.0  \n",
       "8     6       31271394  chr16       2138198.0  \n",
       "9     6       31271394    NaN             NaN  \n",
       "10    8       19956018    NaN             NaN  \n",
       "11   16        2088197    NaN             NaN  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if df.columns[1] == 'hg38':\n",
    "    # I  need to  convert the coordinates to hg19\n",
    "    # Let's get the coordinates from the input file  \n",
    "    df[\"chro\"] = df[\"hg38\"].apply(lambda x: x.split(\":\")[0])\n",
    "    df[\"location_hg38\"] = df[\"hg38\"].apply(lambda x: x.split(\":\")[1])\n",
    "    df['location_hg38'] = df['location_hg38'].apply(pd.to_numeric)\n",
    "    converter = get_lifter('hg38', 'hg19')\n",
    "    Chrom2Pos_hg19 = dict()\n",
    "    Chrom2Pos_hg38 = dict(zip(df.chro, df.location_hg38))\n",
    "    for index, chro,location_hg38 in zip(df.index,df.chro, df.location_hg38): \n",
    "        Chrom = chro\n",
    "        pos = location_hg38\n",
    "        Chrom2Pos_hg19[converter.query(chro, location_hg38)[0][0]]= converter.query(chro, location_hg38)[0][1] \n",
    "df[['chro_N','location_hg_19']] = pd.DataFrame(list(Chrom2Pos_hg19.items()))\n",
    "df\n",
    "#for_test2 = pd.DataFrame(list(Chrom2Pos_hg19.items()),columns = ['CHROM','POS'])\n",
    "\n",
    "# Now we do and save the convertion \n",
    "#for Chrom,Pos in Chrom2Pos_hg38.items():\n",
    "#    chrom = Chrom\n",
    "#    pos = Pos\n",
    "#    converter[chrom][pos]\n",
    "#    Chrom2Pos_hg19[converter.query(chrom, pos)[0][0]]= converter.query(chrom, pos)[0][1] \n",
    "#    Chrom2Pos_hg38 = dict(zip(for_test.CHROM, for_test.POS))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-110-9635685d0714>, line 6)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-110-9635685d0714>\"\u001b[0;36m, line \u001b[0;32m6\u001b[0m\n\u001b[0;31m    r = requests.get(server+str(chro_N)+':g.'+str(location_hg_19)+)\u001b[0m\n\u001b[0m                                                                  ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "server = 'http://myvariant.info/v1/variant/'\n",
    "\n",
    "for index, chro_N,location_hg_19 in zip(df.index,df.chro_N, df.location_hg_19): \n",
    "    \n",
    "    # Request API here\n",
    "    r = requests.get(server+str(chro_N)+':g.'+str(location_hg_19)+)\n",
    "    \n",
    "    decoded = r.json()\n",
    "    # We select the info we want and store it in dict\n",
    "    \n",
    "    # For row2gene_symbol\n",
    "    row2gene_symbol[index]= decoded['hits'][0]['dbsnp']['gene']['symbol']\n",
    "\n",
    "    # For row2CHROM\n",
    "    #row2CHROM[index] = decoded['hits'][0]['chrom']\n",
    "    \n",
    "    # For row2POS \n",
    "    #row2POS[index] = decoded['hits'][0]['vcf']['position']\n",
    "    \n",
    "    # For row2ALT \n",
    "    #row2ALT[index] = decoded['hits'][0]['vcf']['alt']\n",
    "    \n",
    "    # For row2REF\n",
    "    #row2REF[index] = decoded['hits'][0]['vcf']['ref']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'http://myvariant.info/v1/variant/chr3100951675'"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server+str(chro_N)+str(location_hg_19)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'code': 404, 'success': False, 'error': \"ID 'chr3100951675' not found\"}"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "decoded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "http://myvariant.info/v1/variant/chr1:g.35367G>A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>hg38</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "      <th>chro</th>\n",
       "      <th>location_hg38</th>\n",
       "      <th>chro_N</th>\n",
       "      <th>location_hg_19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>chr3</td>\n",
       "      <td>100951675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>chr17</td>\n",
       "      <td>41219633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>chr15</td>\n",
       "      <td>48892422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>chr14</td>\n",
       "      <td>23896042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>chr9</td>\n",
       "      <td>136131429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117199558</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>7</td>\n",
       "      <td>117199558</td>\n",
       "      <td>chr7</td>\n",
       "      <td>116839612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>7</td>\n",
       "      <td>117559502</td>\n",
       "      <td>chr7</td>\n",
       "      <td>117199556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344XX-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>chr6</td>\n",
       "      <td>31239171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>8</td>\n",
       "      <td>19956018</td>\n",
       "      <td>chr8</td>\n",
       "      <td>19813529</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>16</td>\n",
       "      <td>2088197</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2138198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene         hg38                         Transcript Quality   Filter  \\\n",
       "0   IMPG2  3:101232831              NM_016247.3:c.3183C>G   34/64     PASS   \n",
       "1   BRCA1  17:43067616              NM_007294.3:c.5066T>C  49/131     PASS   \n",
       "2    FBN1  15:48600225               NM_000138.4:c.356G>A   46/99     PASS   \n",
       "3    MYH7  14:23426833              NM_000257.3:c.1988G>A   8/102  NO PASS   \n",
       "4   BORIS  9:133256042                         rs56116432   1/200  NO PASS   \n",
       "5    CFTR  7:117199558  ENST00000003084:c.1431_1433delTTC  99/100     PASS   \n",
       "6    CFTR  7:117559502  ENST00000003084:c.1431_1433delTTC  99/100     PASS   \n",
       "7   HLA-C   6:31271394    NM_002117.5:c.344-46_344-45insT   34/64     PASS   \n",
       "8   HLA-C   6:31271394  NM_002117.5:c.344XX-46_344-45insT   34/64     PASS   \n",
       "9   HLA-C   6:31271394                          rs9281300   34/64     PASS   \n",
       "10    LPL   8:19956018               NM_000237.2:c.953A>G   34/64     PASS   \n",
       "11   TSC2   16:2088197                          rs1799758   34/64     PASS   \n",
       "\n",
       "   chro  location_hg38 chro_N location_hg_19  \n",
       "0     3      101232831   chr3      100951675  \n",
       "1    17       43067616  chr17       41219633  \n",
       "2    15       48600225  chr15       48892422  \n",
       "3    14       23426833  chr14       23896042  \n",
       "4     9      133256042   chr9      136131429  \n",
       "5     7      117199558   chr7      116839612  \n",
       "6     7      117559502   chr7      117199556  \n",
       "7     6       31271394   chr6       31239171  \n",
       "8     6       31271394   chr6       31239171  \n",
       "9     6       31271394   chr6       31239171  \n",
       "10    8       19956018   chr8       19813529  \n",
       "11   16        2088197  chr16        2138198  "
      ]
     },
     "execution_count": 193,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   c1   c2 newColumn\n",
      "0  10  100         5\n",
      "1  11  110         3\n",
      "2  12  120         5\n"
     ]
    }
   ],
   "source": [
    "import random\n",
    "\n",
    "inp = [{'c1':10, 'c2':100}, {'c1':11,'c2':110}, {'c1':12,'c2':120}]\n",
    "df = pd.DataFrame(inp)\n",
    "\n",
    "df['newColumn'] = \"\"\n",
    "yourCondition = True\n",
    "for i in range(len(df)):\n",
    "    # put your condition here\n",
    "    if (yourCondition):\n",
    "        # now you can update what you want\n",
    "        df['newColumn'].values[i] = random.randint(0,9)\n",
    "\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "too many values to unpack (expected 2)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-172-c7b88f8fa2b4>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[0;31m#df.loc[df['>'] == True, 'type'] = 'substitution'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[0;31m#df.loc[df['<'] == True, 'type'] = 'substitution'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 15\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mdf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     16\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     17\u001b[0m    \u001b[0;31m# if df.iloc[index,2].str.contains(\">\"):\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: too many values to unpack (expected 2)"
     ]
    }
   ],
   "source": [
    "# First, let identify what column contains the Variant ID\n",
    "\n",
    "#df['rs'] = df.astype(str).applymap(lambda x: 'rs' in x).any(1)\n",
    "#df['del'] = df.astype(str).applymap(lambda x: 'del' in x).any(1)\n",
    "#df['ins'] = df.astype(str).applymap(lambda x: 'ins' in x).any(1)\n",
    "#df['>'] = df.astype(str).applymap(lambda x: '>' in x).any(1)\n",
    "#df['<'] = df.astype(str).applymap(lambda x: '<' in x).any(1)\n",
    "\n",
    "#df['type']=''\n",
    "#df.loc[df['rs'] == True, 'type'] = 'dbsnp'\n",
    "#df.loc[df['del'] == True, 'type'] = 'deletion'\n",
    "#df.loc[df['ins'] == True, 'type'] = 'insertion'\n",
    "#df.loc[df['>'] == True, 'type'] = 'substitution'\n",
    "#df.loc[df['<'] == True, 'type'] = 'substitution'\n",
    "for index, value in df:\n",
    "    print(index)\n",
    "   # if df.iloc[index,2].str.contains(\">\"):\n",
    "     #   print(value)\n",
    "#df[\"alt\"] = df.iloc[:,2].apply(lambda x: x.split(\">\")[0])\n",
    "#df\n",
    "#df[df['A'].str.contains(\"hello\")]\n",
    "#df.iloc[:,2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
