{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please select genome build for the CSV input file: (a) GRCh37 or (b) GRCh38 b\n",
      "Genome build selected: GRCh38\n",
      "Please select genome build for the VCF file: (a) GRCh37 or (b) GRCh38 a\n",
      "Genome build selected: GRCh37\n",
      "Please provide an email address (this will be used to retrieve data from Entrez)mamnamn@gmail.com\n",
      "You entered: mamnamn@gmail.com\n",
      "errors found - refer to output error report\n",
      "Ta-daa!\n",
      "Have a look in the output folder\n",
      "Your files are ready!\n",
      "chr6:g.31239170_31239171insT\n",
      "chr7:g.117199555_117199558del\n",
      "chr7:g.117199555_117199558del\n"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "# coding: utf-8 \n",
    "\n",
    "# Let's import all modules used in this script:\n",
    "\n",
    "\n",
    "import os\n",
    "import subprocess\n",
    "\n",
    "try:\n",
    "    exec(\"import Cython\")\n",
    "except:\n",
    "    subprocess.call(['pip', 'install', 'Cython'])\n",
    "\n",
    "modules = ['sys', 'requests', 'chardet', 'json', 'pandas', 'numpy', 'Bio', 're', 'liftover', 'jinja2', \"warnings\"]\n",
    "\n",
    "for library in modules:\n",
    "    try:\n",
    "        exec(\"import {module}\".format(module=library))\n",
    "    except:\n",
    "        subprocess.call(['pip', 'install', library])\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import inspect\n",
    "from datetime import date\n",
    "#from pathlib import Path\n",
    "import jinja2\n",
    "# For some reason requests did not work sometimes when testing our script in a new env\n",
    "import requests\n",
    "from liftover import get_lifter\n",
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "# To make clear the interection user-cumputer\n",
    "# Let´s delete one warning \n",
    "warnings.filterwarnings(\"ignore\")\n",
    "#################################################################################\n",
    "\n",
    "# Let's read input\n",
    "# This read the file located in the folder input\n",
    "actual_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))\n",
    "#(os.path.dirname(os.path.realpath('CSVtoVCFv3.py')))\n",
    "folder_input = '/input/'\n",
    "folder_output = '/output/'\n",
    "\n",
    "input_file_name = os.listdir(actual_path+folder_input)\n",
    "input_file_name= ''.join(input_file_name)\n",
    "conv38to19 = get_lifter('hg38', 'hg19')\n",
    "\n",
    "CSV_input = actual_path+folder_input+input_file_name\n",
    "#COVERSHEET_TEMPLATE = actual_path+'/Lib/test_report_template.html'\n",
    "\n",
    "\n",
    "# input to data frame\n",
    "\n",
    "# To avoid UnicodeDecodeError: 'utf-8'\n",
    "with open(CSV_input, 'rb') as f:\n",
    "    result = chardet.detect(f.read())\n",
    "\n",
    "#  This not work 100% of the times. To avoid UnicodeDecodeError: 'utf-8'\n",
    "# We strongly recommend that the input be .csv\n",
    "\n",
    "df = pd.read_csv(CSV_input, encoding=result['encoding'])\n",
    "\n",
    "# We will need a extra copy to avoid repetitions\n",
    "df = df.rename(columns=str.upper)\n",
    "df_draft = df\n",
    "\n",
    "\n",
    "\n",
    "if bool(\"GENE\" in df.columns) == True and bool(\"COORDINATES\" in df.columns) == True:\n",
    "    is_stage2 = 'YES'\n",
    "    # Select input genome build for the CSV file\n",
    "    genome_build = input('Please select genome build for the CSV input file: (a) GRCh37 or (b) GRCh38 ')\n",
    "    while not (genome_build == 'a' or genome_build == 'b'):\n",
    "        print('Genome build not supported.')\n",
    "        genome_build = input('Please select genome build for the CSV input file: (a) GRCh37 or (b) GRCh38 ')\n",
    "    if genome_build == 'a':\n",
    "        print('Genome build selected: GRCh37')\n",
    "    if genome_build == 'b':\n",
    "        print('Genome build selected: GRCh38')\n",
    "else:\n",
    "    is_stage2 = 'NO'\n",
    "\n",
    "\n",
    "# Select which genome build to produce the  VCF file in\n",
    "genome_vcf_out= input('Please select genome build for the VCF file: (a) GRCh37 or (b) GRCh38 ')\n",
    "while not (genome_vcf_out == 'a' or genome_vcf_out == 'b'):\n",
    "    print('Genome build not supported.')\n",
    "    genome_vcf_out = input('Please select genome build for the VCF file: (a) GRCh37 or (b) GRCh38 ')\n",
    "if genome_vcf_out == 'a':\n",
    "    print('Genome build selected: GRCh37')\n",
    "if genome_vcf_out == 'b':\n",
    "    print('Genome build selected: GRCh38')\n",
    "\n",
    "# Input email address to retrieve data from Entrez\n",
    "Entrez_ID = input('Please provide an email address (this will be used to retrieve data from Entrez)')\n",
    "while Entrez_ID.find('@')==-1:\n",
    "    print('Email address not entered correctly')\n",
    "    Entrez_ID = input('Please provide an email address (this will be used to retrieve data from Entrez):')\n",
    "\n",
    "if Entrez_ID.find('@')>0:\n",
    "    print('You entered: ' + Entrez_ID)\n",
    "\n",
    "def get_html_gdrive(ID):\n",
    "    URL = \"https://drive.google.com/uc?export=download\"\n",
    "    session = requests.Session()\n",
    "    response = session.get(URL, params = { 'id' : ID }, stream = True)\n",
    "    token = get_confirm_token(response)\n",
    "    if token:\n",
    "        params = { 'id' : ID, 'confirm' : token }\n",
    "        response = session.get(URL, params = params, stream = True)\n",
    "    #response = response.iter_content(32768)\n",
    "    return re.sub('[\\r\\t]','', response.text)\n",
    "def get_confirm_token(response):\n",
    "    for key, value in response.cookies.items():\n",
    "        if key.startswith('download_warning'):\n",
    "            return value\n",
    "    return None\n",
    "html_template = get_html_gdrive('15djn2ZyaaYm9lIlupQ_9J_od9g4eC_-m')\n",
    "\n",
    "# Reading CSV file values and looking for variants IDs\n",
    "\n",
    "# Our programme recognise e.g AGT:c.803T>C, ENST00000003084:c.1431_1433delTTC and rs17289390)\n",
    "\n",
    "def find_trans(data_frame):\n",
    "    pattern = re.compile('(:c.)|(rs\\d+)')\n",
    "    return data_frame[data_frame.apply(lambda x: x.str.contains(pattern))].dropna(how='all').dropna(axis=1, how='all')\n",
    "\n",
    "Transcript = find_trans(df)\n",
    "\n",
    "# Now, we save the results found in a dict key=index and value=variand ID\n",
    "if Transcript.empty == False:\n",
    "    ind = Transcript.index.to_list()\n",
    "    vals = list(Transcript.stack().values)\n",
    "    row2Transcript = dict(zip(ind, vals))\n",
    "    # We need to remove the row where rs has been found to avoid repetitions\n",
    "    # In case in same row more than one kind of ID Variant is stored\n",
    "    for index, Transcript in row2Transcript.items():\n",
    "        # This will be done in df_draft\n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "#print('Is the DataFrame empty? ', df_draft.empty)\n",
    "# df_draft empty means every row contained a rs or :c. ID\n",
    "# print('Is the DataFrame empty? ', df_draft.empty)\n",
    "\n",
    "\n",
    "\n",
    "def get_API(transcripts, data_frame):\n",
    "    df2 = data_frame\n",
    "    df2['exist'] = ''\n",
    "    trans_list = df2['exist']\n",
    "    trans_list = trans_list.to_frame()\n",
    "    trans_list.columns = ['ID']\n",
    "    decoded = dict()\n",
    "    server1 = 'https://rest.ensembl.org/vep/human/hgvs/'\n",
    "    server2 = \"https://rest.ensembl.org/vep/human/id/\"\n",
    "    server_end = '/expand=1;content-type=application/json'\n",
    "    interrogation = '?'\n",
    "    for row, transcript in transcripts.items():\n",
    "        trans1 = ':c.'\n",
    "        trans2 = 'rs'\n",
    "        if trans1 in transcript:\n",
    "            #r = requests.get(server1 + transcript, headers={\"Content-Type\": \"application/json\"})\n",
    "            r = requests.get(server1 + transcript + interrogation + server_end)\n",
    "            if r.status_code == 200:\n",
    "                decoded[row] = r.json()\n",
    "                df2['exist'][row] = 'YES'\n",
    "                trans_list['ID'][row] = transcript\n",
    "            else:\n",
    "                df2['exist'][row] = 'NO'\n",
    "        elif trans2 in transcript:\n",
    "            #r = requests.get(server2 + transcript + interrogation, headers={\"Content-Type\": \"application/json\"})\n",
    "            r = requests.get(server2 + transcript + interrogation + server_end)\n",
    "            if r.status_code == 200:\n",
    "                decoded[row] = r.json()\n",
    "                df2['exist'][row] = 'YES'\n",
    "                trans_list['ID'][row] = transcript\n",
    "            else:\n",
    "                df2['exist'][row] = 'NO'\n",
    "        else:\n",
    "            print(\"unknown transcript\")\n",
    "    trans_df = trans_list[~trans_list.ID.str.contains('^$')]\n",
    "    return decoded, df2, trans_df\n",
    "\n",
    "decoded, df_exist, trans_df = get_API(row2Transcript, df)\n",
    "#trans_df2 = trans_df[~trans_df.ID.str.contains('^$')]\n",
    "##### GENERATE Table with API request errors ####\n",
    "\n",
    "#### This function adds colour highlights to rows where variant has been succesfully retrievd from the API or not\n",
    "def highlight_error(s, num_columns):\n",
    "    if s.exist == 'YES':\n",
    "        return ['background-color: #c0ffba']*num_columns\n",
    "    elif s.exist == 'NO':\n",
    "        return ['background-color: red']*num_columns\n",
    "    elif s.exist == '':\n",
    "        return ['background-color: orange']*num_columns\n",
    "    else:\n",
    "        return ['background-color: blue']*num_columns\n",
    "\n",
    "style1 = df_exist.style.apply(highlight_error, num_columns=len(df_exist.columns), axis=1)\n",
    "df_html = style1.render()\n",
    "\n",
    "# This function adds the data frame to the html template\n",
    "def generate_report(html_results_table, html_template):\n",
    "    todays_date = date.today().strftime('%d/%m/%Y')\n",
    "    html_text = jinja2.Template(html_template)\n",
    "    #html_text = html_text.render(todays_date=todays_date)\n",
    "    elements = html_results_table.split('</style><table')\n",
    "    width='</style><table class=results'\n",
    "    new_html_table=elements[0]+width+elements[1]\n",
    "    html_text = html_text.render(todays_date=todays_date, content=new_html_table)\n",
    "    #html_text += new_html_table\n",
    "    #html_text += '</div>'\n",
    "    return html_text\n",
    "html_file = generate_report(df_html, html_template)\n",
    "\n",
    "# Let's add gene symbol, chromosome, genome location, reference and alteration\n",
    "#pattern2 = re.compile('(NO)|(^$)')\n",
    "#df_exist2 = df_exist[~df_exist.exist.str.contains(pattern2)]\n",
    "def  get_location(decoded, transcripts):\n",
    "    column_names = ['CHROM', 'POS_START', 'POS_END', 'ID']\n",
    "    df_location = pd.DataFrame(columns=column_names)\n",
    "    for i in decoded:\n",
    "        trans = decoded[i]\n",
    "        df_location.loc[i, 'CHROM'] = trans[0][\"seq_region_name\"]\n",
    "        start = trans[0][\"start\"]\n",
    "        end = trans[0][\"end\"]\n",
    "        if start > end:\n",
    "            df_location.loc[i, 'POS_START'] = end\n",
    "            df_location.loc[i, 'POS_END'] = start\n",
    "        else:\n",
    "            df_location.loc[i, 'POS_START'] = start\n",
    "            df_location.loc[i, 'POS_END'] = end\n",
    "        df_location.loc[i, 'CHROM'] = trans[0][\"seq_region_name\"]\n",
    "\n",
    "        df_location.loc[i, 'ID'] = transcripts.loc[i, 'ID']\n",
    "    return df_location\n",
    "\n",
    "df_location = get_location(decoded, trans_df)\n",
    "\n",
    "     # Always tell NCBI who you are\n",
    "\n",
    "# Call Entrez to retrieve sequence in fasta format\n",
    "def call_Entrez(chrom_ID, start, end, Entrez_ID):\n",
    "    Entrez.email = Entrez_ID\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=chrom_ID, rettype=\"fasta\",strand=1,seq_start=start,seq_stop=end)\n",
    "    record = SeqIO.read(handle, \"fasta\")\n",
    "    handle.close()\n",
    "    return str(record.seq)\n",
    "\n",
    "def stage1(decoded, df_location, Entrez_ID):\n",
    "    chromosomes = {\"1\": \"CM000663.2\", \"2\": \"CM000664.2\", \"3\": \"CM000665.2\", \"4\": \"CM000666.2\",\n",
    "                   \"5\": \"CM000667.2\", \"6\": \"CM000668.2\", \"7\": \"CM000669.2\", \"8\": \"CM000670.2\",\n",
    "                   \"9\": \"CM000671.2\", \"10\": \"CM000672.2\", \"11\": \"CM000673.2\", \"12\": \"CM000674.2\",\n",
    "                   \"13\": \"CM000675.2\", \"14\": \"CM000676.2\", \"15\": \"CM000677.2\", \"16\": \"CM000678.2\",\n",
    "                   \"17\": \"CM000679.2\", \"18\": \"CM000680.2\", \"19\": \"CM000681.2\", \"20\": \"CM000682.2\",\n",
    "                   \"21\": \"CM000683.2\", \"22\": \"CM000684.2\", \"X\": \"CM000685.2\", \"Y\": \"CM000686.2\",\n",
    "                   \"M\": \"MF737176\", \"MT\": \"MF737176\"}\n",
    "    #column_names = [\"Gene_API\", 'CHROM', 'POS_HG38', 'POS_HG19', 'ID', 'REF', 'ALT', 'STRAND', 'QUAL', 'FILTER', 'INFO']\n",
    "    column_names = [\"Gene_API\", 'CHROM', 'POS_HG38', 'POS_HG19', 'ID', 'REF', 'ALT', 'STRAND']\n",
    "    df_stage1 = pd.DataFrame(columns=column_names)\n",
    "    # extract specific data from API and data frame into new data frame\n",
    "    for i in decoded:\n",
    "        trans = decoded[i]\n",
    "        # For Gene_symbol\n",
    "        try:\n",
    "            df_stage1.loc[i, 'Gene_API'] = trans[0][\"transcript_consequences\"][0]['gene_symbol']\n",
    "        except KeyError:\n",
    "            df_stage1.loc[i, 'Gene_API'] = '.'\n",
    "        # For CHROMOSOME\n",
    "        df_stage1.loc[i,'CHROM'] = trans[0][\"seq_region_name\"]\n",
    "        # For transcript\n",
    "        df_stage1.loc[i, 'ID'] = df_location.loc[i, 'ID']\n",
    "        # For POSITION in GRCh38\n",
    "        start = trans[0][\"start\"]\n",
    "        end = trans[0][\"end\"]\n",
    "        if start > end:\n",
    "            start = end\n",
    "        # Strand\n",
    "        df_stage1.loc[i, 'STRAND'] = trans[0]['strand']\n",
    "        # REF and ALT on +tive stand\n",
    "        if df_stage1.loc[i, 'STRAND'] == 1:\n",
    "            # For REFERENCE SEQUENCE\n",
    "            ref1 = trans[0]['allele_string'].split('/')[0]\n",
    "            alt1 = trans[0]['allele_string'].split('/')[1]\n",
    "            if ref1 == '-':\n",
    "                ref1_seq = call_Entrez(chromosomes[df_location.loc[i, 'CHROM']], df_location.loc[i, 'POS_START'],\n",
    "                                       df_location.loc[i, 'POS_START'], Entrez_ID)\n",
    "                df_stage1.loc[i,'REF'] = ref1_seq\n",
    "                df_stage1.loc[i,'ALT'] = ref1_seq + alt1\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start\n",
    "            elif alt1 == '-':\n",
    "                ref2_seq = call_Entrez(chromosomes[df_location.loc[i, 'CHROM']], (df_location.loc[i, 'POS_START']-1),\n",
    "                                       df_location.loc[i, 'POS_END'], Entrez_ID)\n",
    "                df_stage1.loc[i, 'REF'] = ref2_seq\n",
    "                df_stage1.loc[i, 'ALT'] = ref2_seq[0]\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start-1\n",
    "                #update_position = conv38to19[df_location.loc[i, 'CHROM']][(df_location.loc[i, 'POS_START'] - 1)]\n",
    "                #df_stage1.loc[i, 'POS_HG19'] = update_position[0][1]\n",
    "            else:\n",
    "                df_stage1.loc[i, 'REF'] = ref1\n",
    "                df_stage1.loc[i, 'ALT'] = alt1\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start\n",
    "            df_stage1.loc[i, 'STRAND'] = '+'\n",
    "        # REF and ALT on -tive strand\n",
    "        if df_stage1.loc[i, 'STRAND'] == -1:\n",
    "            ref2 = trans[0]['allele_string'].split('/')[0]\n",
    "            alt2 = trans[0]['allele_string'].split('/')[1]\n",
    "            complement = {'A': 'T', 'G': 'C', 'C': 'G', 'T': 'A', '-': '-'}\n",
    "            # For REFERENCE SEQUENCE\n",
    "            ref3 = \"\".join(complement[c] for c in ref2)[::-1]\n",
    "            # For Variant Sequence\n",
    "            alt3 = \"\".join(complement[d] for d in alt2)[::-1]\n",
    "            if ref2 == '-':\n",
    "                ref3_seq = call_Entrez(chromosomes[df_location.loc[i, 'CHROM']], df_location.loc[i, 'POS_START'],\n",
    "                                       df_location.loc[i, 'POS_START'], Entrez_ID)\n",
    "                df_stage1.loc[i, 'REF'] = ref3_seq\n",
    "                df_stage1.loc[i, 'ALT'] = ref3_seq + alt3\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start\n",
    "            elif alt2 == '-':\n",
    "                ref4_seq = call_Entrez(chromosomes[df_location.loc[i, 'CHROM']], (df_location.loc[i, 'POS_START']-1),\n",
    "                                       df_location.loc[i, 'POS_END'], Entrez_ID)\n",
    "                df_stage1.loc[i, 'REF'] = ref4_seq\n",
    "                df_stage1.loc[i, 'ALT'] = ref4_seq[0]\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start - 1\n",
    "                #update_position2 = conv38to19[df_location.loc[i, 'CHROM']][(df_location.loc[i, 'POS_START'] + 1)]\n",
    "                #df_stage1.loc[i, 'POS_HG19'] = update_position2[0][1]\n",
    "            else:\n",
    "                df_stage1.loc[i, 'REF'] = ref3\n",
    "                df_stage1.loc[i, 'ALT'] = alt3\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start\n",
    "            # For REFERENCE SEQUENCE\n",
    "            #df_stage1.loc[i, 'REF'] = \"\".join(complement[c] for c in ref2)[::-1]\n",
    "            # For Variant Sequence\n",
    "            #df_stage1.loc[i, 'ALT'] = \"\".join(complement[d] for d in alt2)[::-1]\n",
    "            df_stage1.loc[i, 'STRAND'] = '-'\n",
    "        # For ID\n",
    "\n",
    "    #df_stage1['ID'] = data_frame[[\"Transcript\"]].copy()\n",
    "        # For Quality\n",
    "    #df_stage1['QUAL'] = data_frame[[\"Quality\"]].copy()\n",
    "        # For Quality\n",
    "    #df_stage1['FILTER'] = data_frame[[\"Filter\"]].copy()\n",
    "    return df_stage1\n",
    "\n",
    "\n",
    "df_stage1 = stage1(decoded, df_location, Entrez_ID)\n",
    "\n",
    "def get_hg19(df):\n",
    "    for index, row in df.iterrows():\n",
    "        chrom = row['CHROM']\n",
    "        if chrom == 'MT':\n",
    "            chrom = 'M'\n",
    "        pos = row['POS_HG38']\n",
    "        converted = conv38to19[chrom][pos]\n",
    "        row['POS_HG19'] = converted[0][1]\n",
    "    return df\n",
    "\n",
    "df_stage1 = get_hg19(df_stage1)\n",
    "\n",
    "\n",
    "############ End of the STAGE 1 ######################\n",
    "\n",
    "\n",
    "# Stage 2\n",
    "\n",
    "# First, let's prepare where our information file is going to be create.\n",
    "\n",
    "# Introce in path the directory where we wish to create the new CVF\n",
    "# save_path = '/Users/manolodominguez/Desktop/git-repos/STP_mini_projects/Igor-Manuel/Outputs'\n",
    "# save_path = '/Users/Igor/Google Drive/Clinical_Bioinformatics_STP/Igor-Manuel_Project/Outputs'\n",
    "# Introduce in name, the name of your new file\n",
    "\n",
    "\n",
    "txt_file = \"CSVtoVCF_ErrorReport.html\"\n",
    "\n",
    "output_inform = actual_path + folder_output + txt_file\n",
    "\n",
    "# This will  create directory + file name\n",
    "# completeName = os.path.join(save_path, name_of_file + \".txt\")\n",
    "\n",
    "# To compare input with API result\n",
    "# We need to create a new column from CHROM + ':' + POS\n",
    "\n",
    "#df_stage1 = df_stage1.applymap(str)\n",
    "\n",
    "def get_genome(data_frame, genome):\n",
    "    df_stage1 = data_frame.applymap(str)\n",
    "    if genome == 'a':\n",
    "        df_stage1['Location'] = df_stage1['CHROM'].str.cat(df_stage1['POS_HG19'], sep=':')\n",
    "    if genome == 'b':\n",
    "        df_stage1['Location'] = df_stage1['CHROM'].str.cat(df_stage1['POS_HG38'], sep=':')\n",
    "    return df_stage1\n",
    "\n",
    "def get_stage2(df_stage1, df_filtered, genome):\n",
    "    df_stage1_final = get_genome(df_stage1, genome)\n",
    "#df_stage1['Location'] = df_stage1['CHROM'].str.cat(df_stage1['POS'], sep=':')\n",
    "\n",
    "    df_comparation = pd.concat([df_filtered[['GENE', 'COORDINATES']], df_stage1_final[['ID', 'Gene_API', 'Location']]], axis=1)\n",
    "# df_comparation = df_comparation.set_index(\"Transcript\")\n",
    "\n",
    "# New two columns as result of the comparation\n",
    "    df_comparation['result1'] = np.where(df_comparation.iloc[:, 0] == df_comparation.iloc[:, 3], 'OK', 'ERROR')\n",
    "\n",
    "    df_comparation['result2'] = np.where(df_comparation.iloc[:, 1] == df_comparation.iloc[:, 4], 'OK', 'ERROR')\n",
    "\n",
    "# We save here the rows where errors in gene symbol have been found\n",
    "    gene_error = df_comparation[df_comparation['result1'].str.match('ERROR')]\n",
    "    gene_error = gene_error.drop(gene_error.columns[[1, 4, 5, 6]], axis=1)\n",
    "\n",
    "# We save here the rows where errors in location have been found\n",
    "    location_error = df_comparation[df_comparation['result2'].str.match('ERROR')]\n",
    "    location_error = location_error.drop(location_error.columns[[0, 3, 5, 6]], axis=1)\n",
    "    return gene_error, location_error\n",
    "\n",
    "def gene_loc_error(html_file, location_error, gene_error):\n",
    "    variant2locations = location_error.set_index('ID').T.to_dict('list')\n",
    "    # we save this information in a dict with the variant where the error has been found as key\n",
    "    # Plus in a list the gene simbol of the input and the gene simbol of the API as value.\n",
    "    gene2gene_API = gene_error.set_index('ID').T.to_dict('list')\n",
    "\n",
    "    # If there is not any mismatch found, one document will be created saying\n",
    "    No_error_found = 'Congratulation, no errors have been found in your CSV with regard to gene symbol and location\\n'\n",
    "    # If there are mismathes found, one document will be created saying\n",
    "    Errors_found = 'If you are reading this message, then the CSVtoVCF application detected some error(s) or mismatch(s) in the Gene Symbol and/or the location of your variants as compared to data retrieved from API Emsembl (VEP).\\n Below the differences found are shown.\\n\\n'\n",
    "    # Plus the mismatch\n",
    "\n",
    "    # Number of locations mismatches\n",
    "    how_many_location_error = 'Number of location mismatch(s) found:', str(len(location_error))\n",
    "    how_many_location_error = ' '.join(how_many_location_error)\n",
    "\n",
    "\n",
    "    # Now, show these mismatchs in this order\n",
    "    header_location = '<th class=\"col_heading level0 col0\" >Your location</th> <th class=\"col_heading level0 col1\" >Variant</th>  <th class=\"col_heading level0 col2\" >API location</th>'\n",
    "    html = html_file\n",
    "\n",
    "# Now with these dicts we can write the inform_error.txt\n",
    "# If both dicts are empty\n",
    "# then create the txt saying there is not error found\n",
    "# If one or both dict contain data\n",
    "# that means mismatch found in the input\n",
    "\n",
    "# Now the same with gene_symbol\n",
    "\n",
    "# Number of gene_symbol mismatchs\n",
    "    how_many_GS_error = 'Number of Gene Symbol mismatch(s) found:', str(len(gene_error))\n",
    "    how_many_GS_error = ' '.join(how_many_GS_error)\n",
    "\n",
    "# Now, show these mismatchs in this order\n",
    "\n",
    "    header_gene_symbol = '<th class=\"col_heading level0 col0\" >Your Gene_symbol</th> <th class=\"col_heading level0 col1\" >Variant</th>  <th class=\"col_heading level0 col2\" >API Gene_symbol</th>'\n",
    "\n",
    "\n",
    "    location_title = \"############# LOCATION'S MISMATCH(S) ############## \\n\\n\"\n",
    "    gene_title = \"\\n\\n############# GENE_SYMBOL'S MISMATCH(S) #############\\n\\n\"\n",
    "\n",
    "# Here we introduce a condition, if dicts are empty\n",
    "# write No_error_found\n",
    "# if dicts are not empty, write all this information\n",
    "\n",
    "\n",
    "    if (bool(variant2locations) == False) and (bool(gene2gene_API) == False):\n",
    "        html += '<div class=errors_summary><b>\\n'+ No_error_found +'</b></div>\\n'\n",
    "        html += \"</body>\"\n",
    "        html += \"</html>\"\n",
    "    #    with open(output_inform, 'w') as out:\n",
    "    #        out.writelines([No_error_found])\n",
    "    else:\n",
    "        print(\"errors found - refer to output error report\")\n",
    "    # Write location\n",
    "\n",
    "    #with open(output_inform, 'w') as out:\n",
    "    #    out.writelines([Errors_found, location_title, how_many_location_error, separator, header_location])\n",
    "    #location_error.to_csv(output_inform,\n",
    "    #                      header=None, index=None, sep='\\t', mode='a')\n",
    "\n",
    "    # Now, with Gene_symbol\n",
    "\n",
    "    #with open(output_inform, 'a') as out:\n",
    "    #    out.writelines([gene_title, how_many_GS_error, separator, header_gene_symbol])\n",
    "    #gene_error.to_csv(output_inform,\n",
    "    #                  header=None, index=None, sep='\\t', mode='a')\n",
    "        html += '<div class=errors_summary>\\n'\n",
    "        html += '<p>'+Errors_found+'</p>'\n",
    "        html += '<div class=error_sum1>\\n'\n",
    "        html += '<p style=\"text-align:center\">'+ gene_title +'</p>\\n'\n",
    "        html += '<p style=\"text-align:center\">'+ how_many_GS_error +'</p>\\n'\n",
    "        html += '<table class=errors>\\n'+ header_gene_symbol\n",
    "        for j in range(len(gene_error)):\n",
    "            html+= '<tr>\\n'\n",
    "            for col2 in gene_error.columns:\n",
    "                value2 = gene_error.iloc[j][col2]\n",
    "                html +='<td>\\n'+str(value2)+'\\n</td>\\n'\n",
    "            html+= '</tr>\\n'\n",
    "        html +='</table>\\n'\n",
    "        html += '</div>\\n'\n",
    "        html += '<div class=error_sum2>\\n'\n",
    "        html += '<p style=\"text-align:center\">'+ location_title +'</p>\\n'\n",
    "        html += '<p style=\"text-align:center\">'+ how_many_location_error +'</p>\\n'\n",
    "        html += '<table class=errors>\\n'+ header_location\n",
    "        for i in range(len(location_error)):\n",
    "            html+= '<tr>\\n'\n",
    "            for col in location_error.columns:\n",
    "                value = location_error.iloc[i][col]    \n",
    "                html +='<td>\\n'+str(value)+'\\n</td>\\n'\n",
    "            html+= '</tr>\\n'\n",
    "        html +='</table>\\n'\n",
    "        html += '</div>\\n'\n",
    "        html += '</div>\\n'\n",
    "        html += '</body>\\n'\n",
    "        html += \"</html>\"\n",
    "    return html\n",
    "\n",
    "if is_stage2 == 'YES':\n",
    "    pattern = re.compile('(NO)|(^$)')\n",
    "    df_filtered = df_exist[~df_exist.exist.str.contains(pattern)]\n",
    "    gene_error, location_error = get_stage2(df_stage1, df_filtered, genome_build)\n",
    "    html_final = gene_loc_error(html_file, location_error, gene_error)\n",
    "    with open (output_inform, 'w') as out:\n",
    "        out.write(html_final)\n",
    "\n",
    "#  End of stage 2\n",
    "\n",
    "#################### Skip stage two #######################\n",
    "\n",
    "# get error report as html file html file\n",
    "def finish_html(html_file):\n",
    "    text = \"End of error report. CSVtoVCF can also check if your GENE symbol and chromosome \" \\\n",
    "           \"coordinates match with the GENE and location retrieved by API. \" \\\n",
    "           \"To enable this function make sure that your gene names are located under a column with a title 'GENE'\" \\\n",
    "           \"and chromosome coordinates matches example formal (e.g. 17:656767), the column title for this must be 'COORDINATES'.\"\n",
    "    html_file += '<div class=errors_summary><b>\\n'+ text +'</b></div>\\n'\n",
    "    html_file += \"</body>\"\n",
    "    html_file += \"</html>\"\n",
    "    return html_file\n",
    "if is_stage2 == 'NO':\n",
    "    finished_html = finish_html(html_file)\n",
    "    with open (output_inform, 'w') as out:\n",
    "        out.write(finished_html)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# # STAGE 3\n",
    "\n",
    "# The VCF create for us follow the  structure explained in https://samtools.github.io/hts-specs/VCFv4.2.pdf\n",
    "# We have created the minimum structure that a VCF document must have\n",
    "# according to documentation (VCFv4.3 format)\n",
    "\n",
    "# An example here\n",
    "\n",
    "#      ##fileformat=VCFv4.3\n",
    "#      #CHROM POS      ID         REF   ALT    QUAL  FILTER   INFO\n",
    "#      20     14370    rs6054257  G     A      29    PASS    NS=3;DP=14;AF=0.5;DB;H2\n",
    "\n",
    "# Here we explain how we are going to fill the columns of the  VCF file.\n",
    "\n",
    "### Mandatory columns ###\n",
    "# CHROM taken from df_stage1['CHROM']\n",
    "# POS taken from df_stage1['POS']\n",
    "# ID taken df_stage1['ID']\n",
    "# REF taken from df_stage1['REF']\n",
    "# ALT taken from df_stage1['ALT']\n",
    "# QUAL taken from df_stage1['QUAL']\n",
    "# FILTER taken from df_stage1['FILTER']\n",
    "# INFO filled with (‘.’)\n",
    "\n",
    "# First, let's create a empty document in the same directory\n",
    "\n",
    "# Introce in path the directory where we wish to create the new CVF\n",
    "# save_path = '/Users/manolodominguez/Desktop/git-repos/STP_mini_projects/Igor-Manuel/Outputs'\n",
    "\n",
    "# Select which genome build to produce the  VCF file in\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Introduce in name, the name of your new file\n",
    "name_of_VCF = \"your_\" + input_file_name + \"_converted_into_VCF.vcf\"\n",
    "\n",
    "output_inform = actual_path + folder_output + name_of_VCF\n",
    "\n",
    "# This will  create directory + file name\n",
    "# completeName2 = os.path.join(save_path, name_of_file2 + \".vcf\")\n",
    "\n",
    "# Now, let's create the metadata information\n",
    "file_format = '##fileformat=VCFv4.3\\n'\n",
    "today = date.today()\n",
    "d1 = today.strftime(\"%Y-%m-%d\")\n",
    "file_Date = '##fileDate=' + d1 + '\\n'\n",
    "\n",
    "# And now the header of the columns\n",
    "#header_line = '#CHROM\\tPOS\\tID\\tREF\\tALT\\tSTRAND\\tQUAL\\tFILTER\\tINFO\\n'\n",
    "header_line = '#CHROM\\tPOS\\tID\\tREF\\tALT\\tSTRAND\\tGENE\\tINFO\\n'\n",
    "\n",
    "# Now, we write meta-information line and header line\n",
    "with open(output_inform, 'w') as out:\n",
    "    out.writelines([file_format, file_Date, header_line])\n",
    "\n",
    "# Let's modify the df_stage1 to be added to the VCF file\n",
    "df_stage2 = df_stage1\n",
    "df_stage2['INFO'] = '.'\n",
    "\n",
    "def convert(v):\n",
    "    try:\n",
    "        return int(v)\n",
    "    except ValueError:\n",
    "        return v\n",
    "\n",
    "\n",
    "def get_vcf_genome(data_frame, genome_build):\n",
    "    choice = {\"a\": \"POS_HG19\", \"b\": \"POS_HG38\"}\n",
    "    data_frame2 = pd.DataFrame([convert(c) for c in l] for l in data_frame.values).sort_values([1, 2], ascending=(True, True))\n",
    "    data_frame2.columns = data_frame.columns\n",
    "    data_frame2 = data_frame2.sort_values(['CHROM', choice[genome_build]])\n",
    "    #data_frame['CHROM'] = data_frame['CHROM'].astype(float)\n",
    "    #data_frame[choice[genome_build]] = data_frame[choice[genome_build]].astype(float)\n",
    "    #data_frame = data_frame.sort_values([ 'CHROM', choice[genome_build]], ascending=(True, True))\n",
    "    #data_frame = data_frame[['CHROM', choice[genome_build], 'ID', 'REF', 'ALT', 'STRAND', 'QUAL', 'FILTER', 'INFO', 'Gene_API']]\n",
    "    data_frame2 = data_frame2[\n",
    "        ['CHROM', choice[genome_build], 'ID', 'REF', 'ALT', 'STRAND', 'Gene_API', 'INFO']]\n",
    "    data_frame2 = data_frame2.replace(to_replace=\"nan\",\n",
    "                              value=\".\")\n",
    "    return data_frame2\n",
    "\n",
    "df_stage3 = get_vcf_genome(df_stage2, genome_vcf_out)\n",
    "\n",
    "df_stage3.to_csv(output_inform,\n",
    "                 header=None, index=None, sep='\\t', mode='a')\n",
    "\n",
    "print('Ta-daa!')\n",
    "print('Have a look in the output folder')\n",
    "print('Your files are ready!')\n",
    "\n",
    "\n",
    "for idx, value in df_stage3.iloc[:,2].iteritems():\n",
    "    if \">\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + df_stage3.loc[idx, [\"REF\"]][0] + \">\" + df_stage3.loc[idx, [\"ALT\"]][0]\n",
    "    if \"ins\" in value: \n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + \"_\" + ((len(df_stage3.loc[idx,[\"ID\"]].apply(lambda x: x.split(\"ins\")[1])[0]) + df_stage3.loc[idx, [\"POS_HG19\"]][0]).astype(str)) + \"ins\" + df_stage3.loc[8,[\"ID\"]].apply(lambda x: x.split(\"ins\")[1])[0]\n",
    "    if \"del\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + \"_\" + ((len(df_stage3.loc[idx,[\"ID\"]].apply(lambda x: x.split(\"del\"))[0][1]) + df_stage3.loc[idx, [\"POS_HG19\"]][0]).astype(str)) + \"del\" \n",
    "    if \"rs\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = df_stage3.loc[idx, [\"ID\"]][0]\n",
    "        \n",
    "        \n",
    "        \n",
    "server_rs = \"http://myvariant.info/v1/query?q=\"\n",
    "server = 'http://myvariant.info/v1/variant/'\n",
    "\n",
    "for idx, value in df_stage3.iloc[:,8].iteritems():\n",
    "    if ':g.' in value and \">\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            print(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "   \n",
    "    if \"rs\" in value:\n",
    "        r = requests.get(server_rs+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            print(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['hits'][0]['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['hits'][0]['vcf']['alt']\n",
    "            \n",
    "    if ':g.' in value and \"del\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            print(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "    \n",
    "    if ':g.' in value and \"ins\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            print(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "        \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG19</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "      <th>TEST</th>\n",
       "      <th>REF_VariantINFO</th>\n",
       "      <th>ALT_VariantINFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>100951675</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "      <td>chr3:g.100951675G&gt;C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31239170</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.31239170_31239171insT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35427531</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.35427531T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555_117199558del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555_117199558del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>136131429</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23896042</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "      <td>chr14:g.23896042C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48892422</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:g.48892422C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>41219633</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:g.41219633A&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>41223094</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG19                                 ID   REF ALT STRAND  \\\n",
       "0      3  100951675              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31239170    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35427531          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  136131429                         rs56116432     C   T      +   \n",
       "3     14   23896042              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48892422               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   41219633              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   41223094                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API INFO                           TEST REF_VariantINFO ALT_VariantINFO  \n",
       "0    IMPG2    .            chr3:g.100951675G>C               G               C  \n",
       "8    HLA-C    .   chr6:g.31239170_31239171insT             NaN             NaN  \n",
       "9    FANCE    .             chr6:g.35427531T>C               T               C  \n",
       "6     CFTR    .  chr7:g.117199555_117199558del             NaN             NaN  \n",
       "7     CFTR    .  chr7:g.117199555_117199558del             NaN             NaN  \n",
       "4      ABO    .                     rs56116432               C               T  \n",
       "3     MYH7    .            chr14:g.23896042C>T               C               T  \n",
       "2     FBN1    .            chr15:g.48892422C>T               C               T  \n",
       "1    BRCA1    .            chr17:g.41219633A>G               A               G  \n",
       "5    BRCA1    .                      rs1799966               T               A  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stage3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG19</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "      <th>TEST</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>100951675</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "      <td>chr3:g.100951675G&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31239170</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.31239170_31239171insT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35427531</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.35427531T&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555_117199558del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555_117199558del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>136131429</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "      <td>rs56116432</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23896042</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "      <td>chr14:g.23896042C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48892422</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:g.48892422C&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>41219633</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:g.41219633A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>41223094</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>rs1799966</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG19                                 ID   REF ALT STRAND  \\\n",
       "0      3  100951675              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31239170    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35427531          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  136131429                         rs56116432     C   T      +   \n",
       "3     14   23896042              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48892422               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   41219633              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   41223094                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API INFO                           TEST  \n",
       "0    IMPG2    .            chr3:g.100951675G>C  \n",
       "8    HLA-C    .   chr6:g.31239170_31239171insT  \n",
       "9    FANCE    .             chr6:g.35427531T>C  \n",
       "6     CFTR    .  chr7:g.117199555_117199558del  \n",
       "7     CFTR    .  chr7:g.117199555_117199558del  \n",
       "4      ABO    .                     rs56116432  \n",
       "3     MYH7    .            chr14:g.23896042C>T  \n",
       "2     FBN1    .            chr15:g.48892422C>T  \n",
       "1    BRCA1    .            chr17:g.41219633A>G  \n",
       "5    BRCA1    .                      rs1799966  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "for idx, value in df_stage3.iloc[:,2].iteritems():\n",
    "    if \">\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + df_stage3.loc[idx, [\"REF\"]][0] + \">\" + df_stage3.loc[idx, [\"ALT\"]][0]\n",
    "    if \"ins\" in value: \n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + \"_\" + ((len(df_stage3.loc[idx,[\"ID\"]].apply(lambda x: x.split(\"ins\")[1])[0]) + df_stage3.loc[idx, [\"POS_HG19\"]][0]).astype(str)) + \"ins\" + df_stage3.loc[8,[\"ID\"]].apply(lambda x: x.split(\"ins\")[1])[0]\n",
    "    if \"del\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + \"_\" + ((len(df_stage3.loc[idx,[\"ID\"]].apply(lambda x: x.split(\"del\"))[0][1]) + df_stage3.loc[idx, [\"POS_HG19\"]][0]).astype(str)) + \"del\" \n",
    "    if \"rs\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = df_stage3.loc[idx, [\"ID\"]][0]\n",
    "        \n",
    "df_stage3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stage3.loc[idx, [\"REF_VariantINFO\"]] = \"\"\n",
    "df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG19</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "      <th>TEST</th>\n",
       "      <th>REF_VariantINFO</th>\n",
       "      <th>ALT_VariantINFO</th>\n",
       "      <th>test_comparation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>100951675</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "      <td>chr3:g.100951675G&gt;C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31239170</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.31239170_31239171insT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35427531</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.35427531T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555_117199558del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555_117199558del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>136131429</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23896042</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "      <td>chr14:g.23896042C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48892422</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:g.48892422C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>41219633</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:g.41219633A&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>41223094</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG19                                 ID   REF ALT STRAND  \\\n",
       "0      3  100951675              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31239170    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35427531          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  136131429                         rs56116432     C   T      +   \n",
       "3     14   23896042              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48892422               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   41219633              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   41223094                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API INFO                           TEST REF_VariantINFO  \\\n",
       "0    IMPG2    .            chr3:g.100951675G>C               G   \n",
       "8    HLA-C    .   chr6:g.31239170_31239171insT             NaN   \n",
       "9    FANCE    .             chr6:g.35427531T>C               T   \n",
       "6     CFTR    .  chr7:g.117199555_117199558del             NaN   \n",
       "7     CFTR    .  chr7:g.117199555_117199558del             NaN   \n",
       "4      ABO    .                     rs56116432               C   \n",
       "3     MYH7    .            chr14:g.23896042C>T               C   \n",
       "2     FBN1    .            chr15:g.48892422C>T               C   \n",
       "1    BRCA1    .            chr17:g.41219633A>G               A   \n",
       "5    BRCA1    .                      rs1799966               T   \n",
       "\n",
       "  ALT_VariantINFO test_comparation  \n",
       "0               C               OK  \n",
       "8             NaN            ERROR  \n",
       "9               C               OK  \n",
       "6             NaN            ERROR  \n",
       "7             NaN            ERROR  \n",
       "4               T               OK  \n",
       "3               T               OK  \n",
       "2               T               OK  \n",
       "1               G               OK  \n",
       "5               A               OK  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "server_rs = \"http://myvariant.info/v1/query?q=\"\n",
    "server = 'http://myvariant.info/v1/variant/'\n",
    "unknow_variants = []\n",
    "for idx, value in df_stage3.iloc[:,8].iteritems():\n",
    "    if ':g.' in value and \">\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "   \n",
    "    if \"rs\" in value:\n",
    "        r = requests.get(server_rs+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['hits'][0]['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['hits'][0]['vcf']['alt']\n",
    "            \n",
    "    if ':g.' in value and \"del\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "    \n",
    "    if ':g.' in value and \"ins\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "df_stage3\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of variant identify in your file:  7\n",
      "Total number of variants compared successfully:  7\n",
      "Number of variant whose results match in both APIs:  7\n",
      "Number of variant whose results do not match in both APIs:  0\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG19</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "      <th>TEST</th>\n",
       "      <th>REF_VariantINFO</th>\n",
       "      <th>ALT_VariantINFO</th>\n",
       "      <th>test_comparation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>100951675</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "      <td>chr3:g.100951675G&gt;C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35427531</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.35427531T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>136131429</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23896042</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "      <td>chr14:g.23896042C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48892422</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:g.48892422C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>41219633</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:g.41219633A&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>41223094</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG19                         ID REF ALT STRAND Gene_API INFO  \\\n",
       "0      3  100951675      NM_016247.3:c.3183C>G   G   C      -    IMPG2    .   \n",
       "9      6   35427531  ENST00000229769:c.1310T>C   T   C      +    FANCE    .   \n",
       "4      9  136131429                 rs56116432   C   T      +      ABO    .   \n",
       "3     14   23896042      NM_000257.3:c.1988G>A   C   T      -     MYH7    .   \n",
       "2     15   48892422       NM_000138.4:c.356G>A   C   T      -     FBN1    .   \n",
       "1     17   41219633      NM_007294.3:c.5066T>C   A   G      -    BRCA1    .   \n",
       "5     17   41223094                  rs1799966   T   A      +    BRCA1    .   \n",
       "\n",
       "                  TEST REF_VariantINFO ALT_VariantINFO test_comparation  \n",
       "0  chr3:g.100951675G>C               G               C               OK  \n",
       "9   chr6:g.35427531T>C               T               C               OK  \n",
       "4           rs56116432               C               T               OK  \n",
       "3  chr14:g.23896042C>T               C               T               OK  \n",
       "2  chr15:g.48892422C>T               C               T               OK  \n",
       "1  chr17:g.41219633A>G               A               G               OK  \n",
       "5            rs1799966               T               A               OK  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "total_variants = len(df_stage3.index)\n",
    "print(\"Number of variant identify in your file: \",total_variants)\n",
    "unknow_variants = df_stage3[\"REF_VariantINFO\"].isnull().sum(axis=0)\n",
    "#print(\"Number of variants don´t find in the the second API: \",unknow_variants)\n",
    "df_stage3 = df_stage3.dropna(subset=[\"REF_VariantINFO\"])\n",
    "df_stage3[\"test_comparation\"] = np.where(df_stage3[\"REF\"]==df_stage3[\"REF_VariantINFO\"],\"OK\",\"ERROR\")\n",
    "print(\"Total number of variants compared successfully: \",len(df_stage3.index))\n",
    "print(\"Number of variant whose results match in both APIs: \",df_stage3[df_stage3.test_comparation == \"OK\"].shape[0])\n",
    "print(\"Number of variant whose results do not match in both APIs: \",df_stage3[df_stage3.test_comparation == \"ERROR\"].shape[0])\n",
    "print(\"\")\n",
    "df_stage3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "From the total number of  7 variants counted in your VCF file 0 are been not found in the MyVariantINFO API\n",
      "These variants are:  0\n",
      "For the rest the variants found in both APIs the  100.0 % of them match in both APIs.\n"
     ]
    }
   ],
   "source": [
    "total_variants = len(df_stage3.index)\n",
    "unknow_variants = df_stage3[\"REF_VariantINFO\"].isnull().sum(axis=0)\n",
    "print(\"From the total number of \", total_variants, \"variants counted in your VCF file\", unknow_variants, \"are been not found in the MyVariantINFO API\")\n",
    "df_stage3 = df_stage3.dropna(subset=[\"REF_VariantINFO\"])\n",
    "\n",
    "if not unknow_variants:\n",
    "    print(\"These variants are: \", unknow_variants)\n",
    "\n",
    "    \n",
    "df_stage3[\"test_comparation\"] = np.where(df_stage3[\"REF\"]==df_stage3[\"REF_VariantINFO\"],\"OK\",\"ERROR\")\n",
    "ok= df_stage3[df_stage3.test_comparation == \"OK\"].shape[0]\n",
    "print(\"For the rest the variants found in both APIs the \",(total_variants/ok)*100,\"% of them match in both APIs.\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.4285714285714286"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(total_variants/ok)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "total_variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG19</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "      <th>TEST</th>\n",
       "      <th>REF_VariantINFO</th>\n",
       "      <th>ALT_VariantINFO</th>\n",
       "      <th>test_comparation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>100951675</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "      <td>chr3:g.100951675G&gt;C</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31239170</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.31239170_31239171insT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35427531</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.35427531T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555_117199558del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555_117199558del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>136131429</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23896042</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "      <td>chr14:g.23896042C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48892422</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:g.48892422C&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>41219633</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:g.41219633A&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>41223094</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG19                                 ID   REF ALT STRAND  \\\n",
       "0      3  100951675              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31239170    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35427531          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  136131429                         rs56116432     C   T      +   \n",
       "3     14   23896042              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48892422               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   41219633              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   41223094                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API INFO                           TEST REF_VariantINFO  \\\n",
       "0    IMPG2    .            chr3:g.100951675G>C               G   \n",
       "8    HLA-C    .   chr6:g.31239170_31239171insT             NaN   \n",
       "9    FANCE    .             chr6:g.35427531T>C               T   \n",
       "6     CFTR    .  chr7:g.117199555_117199558del             NaN   \n",
       "7     CFTR    .  chr7:g.117199555_117199558del             NaN   \n",
       "4      ABO    .                     rs56116432               C   \n",
       "3     MYH7    .            chr14:g.23896042C>T               C   \n",
       "2     FBN1    .            chr15:g.48892422C>T               C   \n",
       "1    BRCA1    .            chr17:g.41219633A>G               A   \n",
       "5    BRCA1    .                      rs1799966               T   \n",
       "\n",
       "  ALT_VariantINFO test_comparation  \n",
       "0               C               OK  \n",
       "8             NaN            ERROR  \n",
       "9               C               OK  \n",
       "6             NaN            ERROR  \n",
       "7             NaN            ERROR  \n",
       "4               T               OK  \n",
       "3               T               OK  \n",
       "2               T               OK  \n",
       "1               G               OK  \n",
       "5               A               OK  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stage3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
