{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please select genome build for the CSV input file: (a) GRCh37 or (b) GRCh38 a\n",
      "Genome build selected: GRCh37\n",
      "Please select genome build for the VCF file: (a) GRCh37 or (b) GRCh38 a\n",
      "Genome build selected: GRCh37\n",
      "Please provide an email address (this will be used to retrieve data from Entrez)manuel@gmail.com\n",
      "You entered: manuel@gmail.com\n",
      "errors found - refer to output error report\n",
      "Ta-daa!\n",
      "Have a look in the output folder\n",
      "Your files are ready!\n",
      "Running the test...\n",
      "\n",
      "This test takes the results generated in the output VCF performs a new request in an different API. \n",
      " For a variant given If some of its results matched in both APIs, we asummed that this variant is correct.\n",
      "\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'vcf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-74-f5ff149bfc9c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m    664\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstatus_code\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m200\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    665\u001b[0m             \u001b[0mdecoded\u001b[0m \u001b[0;34m=\u001b[0m  \u001b[0mr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 666\u001b[0;31m             \u001b[0mdf_stage3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"REF_VariantINFO\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdecoded\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'hits'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'vcf'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'ref'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    667\u001b[0m             \u001b[0mdf_stage3\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0midx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"ALT_VariantINFO\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdecoded\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'hits'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'vcf'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'alt'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    668\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'vcf'"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/env python\n",
    "# coding: utf-8 \n",
    "\n",
    "\n",
    "# Readme document is available in this link\n",
    "# https://github.com/Manuel-DominguezCBG/Igor-Manuel\n",
    "\n",
    "\n",
    "# Import Modules\n",
    "\"\"\"\n",
    "The idea of this first version is to create a script tha can be used\n",
    "by the costumer by using the terminal. We have adapt this script to install\n",
    "the needed modules by itself.\n",
    "\n",
    "Note: Considering in the future to create a new env with the needed modules\n",
    "intead of install the modules in the computer of the user.\n",
    "\"\"\"\n",
    "import os\n",
    "import subprocess\n",
    "try:\n",
    "    exec(\"import Cython\")\n",
    "except:\n",
    "    subprocess.call(['pip', 'install', 'Cython'])\n",
    "\n",
    "modules = ['sys', 'requests', 'chardet', 'json', 'pandas', 'numpy', 'Bio', 're', 'liftover', 'jinja2', \"warnings\"]\n",
    "\n",
    "for library in modules:\n",
    "    try:\n",
    "        exec(\"import {module}\".format(module=library))\n",
    "    except:\n",
    "        subprocess.call(['pip', 'install', library])\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import inspect\n",
    "from datetime import date\n",
    "import jinja2\n",
    "import requests\n",
    "from liftover import get_lifter\n",
    "from Bio import Entrez, SeqIO\n",
    "\n",
    "\n",
    "# Removing warnings\n",
    "\"\"\"\n",
    "One long warning is generated during the execution, \n",
    "to make the interaction costumer-computer easier\n",
    "we dont show this warning\n",
    "\"\"\"\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# Reading imput\n",
    "\"\"\"\n",
    "The costumer only need to leave the .csv file in the folder \"input\"\n",
    "Output can be found in \"output\" folder.\n",
    "\"\"\"\n",
    "actual_path = os.path.dirname(os.path.abspath(inspect.stack()[0][1]))\n",
    "folder_input = '/input/'\n",
    "folder_output = '/output/'\n",
    "input_file_name = os.listdir(actual_path+folder_input)\n",
    "input_file_name= ''.join(input_file_name)\n",
    "conv38to19 = get_lifter('hg38', 'hg19')\n",
    "CSV_input = actual_path+folder_input+input_file_name\n",
    "\n",
    "\n",
    "# Convertion input into panda data frame\n",
    "\"\"\"\n",
    "In case the costumer donÂ´t save the input in the righ format\n",
    "UnicodeDecodeError: 'utf-8' can occur.\n",
    "Next lines identify the encoding and open in input in the format\n",
    "This doesnt work always. That is the reason we recommend to save the input file\n",
    "in .csv format.\n",
    "\"\"\"\n",
    "with open(CSV_input, 'rb') as f:\n",
    "    result = chardet.detect(f.read())\n",
    "df = pd.read_csv(CSV_input, encoding=result['encoding'])\n",
    "df = df.rename(columns=str.upper)\n",
    "df_draft = df\n",
    "\n",
    "\n",
    "# Gh37 or 38\n",
    "\"\"\"\n",
    "Our script is able to work with GH 37 or 38.\n",
    "In the execution, we ask the user the genome building used\n",
    "We also provide the coordinates results in both buildings.\n",
    "\"\"\"\n",
    "if bool(\"GENE\" in df.columns) == True and bool(\"COORDINATES\" in df.columns) == True:\n",
    "    is_stage2 = 'YES'\n",
    "    # Select input genome build for the CSV file\n",
    "    genome_build = input('Please select genome build for the CSV input file: (a) GRCh37 or (b) GRCh38 ')\n",
    "    while not (genome_build == 'a' or genome_build == 'b'):\n",
    "        print('Genome build not supported.')\n",
    "        genome_build = input('Please select genome build for the CSV input file: (a) GRCh37 or (b) GRCh38 ')\n",
    "    if genome_build == 'a':\n",
    "        print('Genome build selected: GRCh37')\n",
    "    if genome_build == 'b':\n",
    "        print('Genome build selected: GRCh38')\n",
    "else:\n",
    "    is_stage2 = 'NO'\n",
    "\n",
    "\n",
    "# Select which genome build to produce the  VCF file in\n",
    "genome_vcf_out= input('Please select genome build for the VCF file: (a) GRCh37 or (b) GRCh38 ')\n",
    "while not (genome_vcf_out == 'a' or genome_vcf_out == 'b'):\n",
    "    print('Genome build not supported.')\n",
    "    genome_vcf_out = input('Please select genome build for the VCF file: (a) GRCh37 or (b) GRCh38 ')\n",
    "if genome_vcf_out == 'a':\n",
    "    print('Genome build selected: GRCh37')\n",
    "if genome_vcf_out == 'b':\n",
    "    print('Genome build selected: GRCh38')\n",
    "\n",
    "# Input email address to retrieve data from Entrez\n",
    "Entrez_ID = input('Please provide an email address (this will be used to retrieve data from Entrez)')\n",
    "while Entrez_ID.find('@')==-1:\n",
    "    print('Email address not entered correctly')\n",
    "    Entrez_ID = input('Please provide an email address (this will be used to retrieve data from Entrez):')\n",
    "\n",
    "if Entrez_ID.find('@')>0:\n",
    "    print('You entered: ' + Entrez_ID)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# HTML template\n",
    "\"\"\"\n",
    "Netx 2 def are create to get the html templates from google drive\n",
    "\"\"\"\n",
    "def get_html_gdrive(ID):\n",
    "    URL = \"https://drive.google.com/uc?export=download\"\n",
    "    session = requests.Session()\n",
    "    response = session.get(URL, params = { 'id' : ID }, stream = True)\n",
    "    token = get_confirm_token(response)\n",
    "    if token:\n",
    "        params = { 'id' : ID, 'confirm' : token }\n",
    "        response = session.get(URL, params = params, stream = True)\n",
    "    return re.sub('[\\r\\t]','', response.text)\n",
    "def get_confirm_token(response):\n",
    "    for key, value in response.cookies.items():\n",
    "        if key.startswith('download_warning'):\n",
    "            return value\n",
    "    return None\n",
    "html_template = get_html_gdrive('15djn2ZyaaYm9lIlupQ_9J_od9g4eC_-m')\n",
    "\n",
    "# Looking for variants ID in the input\n",
    "\"\"\"\n",
    "Next lines recognise the variants ID and neccesary information\n",
    "from the imput to perform the API request later\n",
    "\"\"\"\n",
    "def find_trans(data_frame):\n",
    "    pattern = re.compile('(:c.)|(rs\\d+)')\n",
    "    return data_frame[data_frame.apply(lambda x: x.str.contains(pattern))].dropna(how='all').dropna(axis=1, how='all')\n",
    "Transcript = find_trans(df)\n",
    "\n",
    "if Transcript.empty == False:\n",
    "    ind = Transcript.index.to_list()\n",
    "    vals = list(Transcript.stack().values)\n",
    "    row2Transcript = dict(zip(ind, vals))\n",
    "    # We need to remove the row where rs has been found to avoid repetitions\n",
    "    # In case in same row more than one kind of ID Variant is stored\n",
    "    for index, Transcript in row2Transcript.items():\n",
    "        # This will be done in df_draft\n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "\n",
    "# API request\n",
    "\"\"\"\n",
    "We made used of Emsembl VEP API\n",
    "to identify variants\n",
    "\"\"\"\n",
    "def get_API(transcripts, data_frame):\n",
    "    df2 = data_frame\n",
    "    df2['exist'] = ''\n",
    "    trans_list = df2['exist']\n",
    "    trans_list = trans_list.to_frame()\n",
    "    trans_list.columns = ['ID']\n",
    "    decoded = dict()\n",
    "    server1 = 'https://rest.ensembl.org/vep/human/hgvs/'\n",
    "    server2 = \"https://rest.ensembl.org/vep/human/id/\"\n",
    "    server_end = '/expand=1;content-type=application/json'\n",
    "    interrogation = '?'\n",
    "    for row, transcript in transcripts.items():\n",
    "        trans1 = ':c.'\n",
    "        trans2 = 'rs'\n",
    "        if trans1 in transcript:\n",
    "            r = requests.get(server1 + transcript + interrogation + server_end)\n",
    "            if r.status_code == 200:\n",
    "                decoded[row] = r.json()\n",
    "                df2['exist'][row] = 'YES'\n",
    "                trans_list['ID'][row] = transcript\n",
    "            else:\n",
    "                df2['exist'][row] = 'NO'\n",
    "        elif trans2 in transcript:\n",
    "            r = requests.get(server2 + transcript + interrogation + server_end)\n",
    "            if r.status_code == 200:\n",
    "                decoded[row] = r.json()\n",
    "                df2['exist'][row] = 'YES'\n",
    "                trans_list['ID'][row] = transcript\n",
    "            else:\n",
    "                df2['exist'][row] = 'NO'\n",
    "        else:\n",
    "            print(\"unknown transcript\")\n",
    "    trans_df = trans_list[~trans_list.ID.str.contains('^$')]\n",
    "    return decoded, df2, trans_df\n",
    "\n",
    "decoded, df_exist, trans_df = get_API(row2Transcript, df)\n",
    "\n",
    "# GENERATE Table with API request errors #\n",
    "\"\"\"\n",
    "We generare a html file in which we inform if variant is found in the API\n",
    "and if the variant is the kind of variant ID we recognise\n",
    "then we compare the gene_symbol and the coordinates and inform the user\n",
    "the results that dont match\n",
    "\"\"\"\n",
    "\n",
    "#### This function adds colour highlights to rows where variant has been succesfully retrievd from the API or not\n",
    "def highlight_error(s, num_columns):\n",
    "    if s.exist == 'YES':\n",
    "        return ['background-color: #c0ffba']*num_columns\n",
    "    elif s.exist == 'NO':\n",
    "        return ['background-color: red']*num_columns\n",
    "    elif s.exist == '':\n",
    "        return ['background-color: orange']*num_columns\n",
    "    else:\n",
    "        return ['background-color: blue']*num_columns\n",
    "\n",
    "style1 = df_exist.style.apply(highlight_error, num_columns=len(df_exist.columns), axis=1)\n",
    "df_html = style1.render()\n",
    "\n",
    "# Generate report\n",
    "\"\"\"\n",
    "This function adds the data frame to the html template\n",
    "\"\"\"\n",
    "def generate_report(html_results_table, html_template):\n",
    "    todays_date = date.today().strftime('%d/%m/%Y')\n",
    "    html_text = jinja2.Template(html_template)\n",
    "    #html_text = html_text.render(todays_date=todays_date)\n",
    "    elements = html_results_table.split('</style><table')\n",
    "    width='</style><table class=results'\n",
    "    new_html_table=elements[0]+width+elements[1]\n",
    "    html_text = html_text.render(todays_date=todays_date, content=new_html_table)\n",
    "    #html_text += new_html_table\n",
    "    #html_text += '</div>'\n",
    "    return html_text\n",
    "html_file = generate_report(df_html, html_template)\n",
    "\n",
    "# Let's add gene symbol, chromosome, genome location, reference and alteration\n",
    "\n",
    "def  get_location(decoded, transcripts):\n",
    "    column_names = ['CHROM', 'POS_START', 'POS_END', 'ID']\n",
    "    df_location = pd.DataFrame(columns=column_names)\n",
    "    for i in decoded:\n",
    "        trans = decoded[i]\n",
    "        df_location.loc[i, 'CHROM'] = trans[0][\"seq_region_name\"]\n",
    "        start = trans[0][\"start\"]\n",
    "        end = trans[0][\"end\"]\n",
    "        if start > end:\n",
    "            df_location.loc[i, 'POS_START'] = end\n",
    "            df_location.loc[i, 'POS_END'] = start\n",
    "        else:\n",
    "            df_location.loc[i, 'POS_START'] = start\n",
    "            df_location.loc[i, 'POS_END'] = end\n",
    "        df_location.loc[i, 'CHROM'] = trans[0][\"seq_region_name\"]\n",
    "        df_location.loc[i, 'ID'] = transcripts.loc[i, 'ID']\n",
    "    return df_location\n",
    "\n",
    "df_location = get_location(decoded, trans_df)\n",
    "\n",
    "\n",
    "# Call Entrez to retrieve sequence in fasta format\n",
    "def call_Entrez(chrom_ID, start, end, Entrez_ID):\n",
    "    Entrez.email = Entrez_ID\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=chrom_ID, rettype=\"fasta\",strand=1,seq_start=start,seq_stop=end)\n",
    "    record = SeqIO.read(handle, \"fasta\")\n",
    "    handle.close()\n",
    "    return str(record.seq)\n",
    "\n",
    "def stage1(decoded, df_location, Entrez_ID):\n",
    "    chromosomes = {\"1\": \"CM000663.2\", \"2\": \"CM000664.2\", \"3\": \"CM000665.2\", \"4\": \"CM000666.2\",\n",
    "                   \"5\": \"CM000667.2\", \"6\": \"CM000668.2\", \"7\": \"CM000669.2\", \"8\": \"CM000670.2\",\n",
    "                   \"9\": \"CM000671.2\", \"10\": \"CM000672.2\", \"11\": \"CM000673.2\", \"12\": \"CM000674.2\",\n",
    "                   \"13\": \"CM000675.2\", \"14\": \"CM000676.2\", \"15\": \"CM000677.2\", \"16\": \"CM000678.2\",\n",
    "                   \"17\": \"CM000679.2\", \"18\": \"CM000680.2\", \"19\": \"CM000681.2\", \"20\": \"CM000682.2\",\n",
    "                   \"21\": \"CM000683.2\", \"22\": \"CM000684.2\", \"X\": \"CM000685.2\", \"Y\": \"CM000686.2\",\n",
    "                   \"M\": \"MF737176\", \"MT\": \"MF737176\"}\n",
    "\n",
    "    column_names = [\"Gene_API\", 'CHROM', 'POS_HG38', 'POS_HG19', 'ID', 'REF', 'ALT', 'STRAND']\n",
    "    df_stage1 = pd.DataFrame(columns=column_names)\n",
    "    # extract specific data from API and data frame into new data frame\n",
    "    for i in decoded:\n",
    "        trans = decoded[i]\n",
    "        # For Gene_symbol\n",
    "        try:\n",
    "            df_stage1.loc[i, 'Gene_API'] = trans[0][\"transcript_consequences\"][0]['gene_symbol']\n",
    "        except KeyError:\n",
    "            df_stage1.loc[i, 'Gene_API'] = '.'\n",
    "        # For CHROMOSOME\n",
    "        df_stage1.loc[i,'CHROM'] = trans[0][\"seq_region_name\"]\n",
    "        # For transcript\n",
    "        df_stage1.loc[i, 'ID'] = df_location.loc[i, 'ID']\n",
    "        # For POSITION in GRCh38\n",
    "        start = trans[0][\"start\"]\n",
    "        end = trans[0][\"end\"]\n",
    "        if start > end:\n",
    "            start = end\n",
    "        # Strand\n",
    "        df_stage1.loc[i, 'STRAND'] = trans[0]['strand']\n",
    "        # REF and ALT on +tive stand\n",
    "        if df_stage1.loc[i, 'STRAND'] == 1:\n",
    "            # For REFERENCE SEQUENCE\n",
    "            ref1 = trans[0]['allele_string'].split('/')[0]\n",
    "            alt1 = trans[0]['allele_string'].split('/')[1]\n",
    "            if ref1 == '-':\n",
    "                ref1_seq = call_Entrez(chromosomes[df_location.loc[i, 'CHROM']], df_location.loc[i, 'POS_START'],\n",
    "                                       df_location.loc[i, 'POS_START'], Entrez_ID)\n",
    "                df_stage1.loc[i,'REF'] = ref1_seq\n",
    "                df_stage1.loc[i,'ALT'] = ref1_seq + alt1\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start\n",
    "            elif alt1 == '-':\n",
    "                ref2_seq = call_Entrez(chromosomes[df_location.loc[i, 'CHROM']], (df_location.loc[i, 'POS_START']-1),\n",
    "                                       df_location.loc[i, 'POS_END'], Entrez_ID)\n",
    "                df_stage1.loc[i, 'REF'] = ref2_seq\n",
    "                df_stage1.loc[i, 'ALT'] = ref2_seq[0]\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start-1\n",
    "                #update_position = conv38to19[df_location.loc[i, 'CHROM']][(df_location.loc[i, 'POS_START'] - 1)]\n",
    "                #df_stage1.loc[i, 'POS_HG19'] = update_position[0][1]\n",
    "            else:\n",
    "                df_stage1.loc[i, 'REF'] = ref1\n",
    "                df_stage1.loc[i, 'ALT'] = alt1\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start\n",
    "            df_stage1.loc[i, 'STRAND'] = '+'\n",
    "        # REF and ALT on -tive strand\n",
    "        if df_stage1.loc[i, 'STRAND'] == -1:\n",
    "            ref2 = trans[0]['allele_string'].split('/')[0]\n",
    "            alt2 = trans[0]['allele_string'].split('/')[1]\n",
    "            complement = {'A': 'T', 'G': 'C', 'C': 'G', 'T': 'A', '-': '-'}\n",
    "            # For REFERENCE SEQUENCE\n",
    "            ref3 = \"\".join(complement[c] for c in ref2)[::-1]\n",
    "            # For Variant Sequence\n",
    "            alt3 = \"\".join(complement[d] for d in alt2)[::-1]\n",
    "            if ref2 == '-':\n",
    "                ref3_seq = call_Entrez(chromosomes[df_location.loc[i, 'CHROM']], df_location.loc[i, 'POS_START'],\n",
    "                                       df_location.loc[i, 'POS_START'], Entrez_ID)\n",
    "                df_stage1.loc[i, 'REF'] = ref3_seq\n",
    "                df_stage1.loc[i, 'ALT'] = ref3_seq + alt3\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start\n",
    "            elif alt2 == '-':\n",
    "                ref4_seq = call_Entrez(chromosomes[df_location.loc[i, 'CHROM']], (df_location.loc[i, 'POS_START']-1),\n",
    "                                       df_location.loc[i, 'POS_END'], Entrez_ID)\n",
    "                df_stage1.loc[i, 'REF'] = ref4_seq\n",
    "                df_stage1.loc[i, 'ALT'] = ref4_seq[0]\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start - 1\n",
    "                \n",
    "            else:\n",
    "                df_stage1.loc[i, 'REF'] = ref3\n",
    "                df_stage1.loc[i, 'ALT'] = alt3\n",
    "                df_stage1.loc[i, 'POS_HG38'] = start\n",
    "            # For REFERENCE SEQUENCE\n",
    "      \n",
    "            df_stage1.loc[i, 'STRAND'] = '-'\n",
    "      \n",
    "    return df_stage1\n",
    "\n",
    "\n",
    "df_stage1 = stage1(decoded, df_location, Entrez_ID)\n",
    "\n",
    "def get_hg19(df):\n",
    "    for index, row in df.iterrows():\n",
    "        chrom = row['CHROM']\n",
    "        if chrom == 'MT':\n",
    "            chrom = 'M'\n",
    "        pos = row['POS_HG38']\n",
    "        converted = conv38to19[chrom][pos]\n",
    "        row['POS_HG19'] = converted[0][1]\n",
    "    return df\n",
    "\n",
    "df_stage1 = get_hg19(df_stage1)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "txt_file = \"CSVtoVCF_ErrorReport.html\"\n",
    "\n",
    "output_inform = actual_path + folder_output + txt_file\n",
    "\n",
    "# To decide the genome building used by the costumer\n",
    "def get_genome(data_frame, genome):\n",
    "    df_stage1 = data_frame.applymap(str)\n",
    "    if genome == 'a':\n",
    "        df_stage1['Location'] = df_stage1['CHROM'].str.cat(df_stage1['POS_HG19'], sep=':')\n",
    "    if genome == 'b':\n",
    "        df_stage1['Location'] = df_stage1['CHROM'].str.cat(df_stage1['POS_HG38'], sep=':')\n",
    "    return df_stage1\n",
    "\n",
    "def get_stage2(df_stage1, df_filtered, genome):\n",
    "    df_stage1_final = get_genome(df_stage1, genome)\n",
    "\n",
    "    df_comparation = pd.concat([df_filtered[['GENE', 'COORDINATES']], df_stage1_final[['ID', 'Gene_API', 'Location']]], axis=1)\n",
    "\n",
    "# New two columns as result of the comparation\n",
    "    df_comparation['result1'] = np.where(df_comparation.iloc[:, 0] == df_comparation.iloc[:, 3], 'OK', 'ERROR')\n",
    "\n",
    "    df_comparation['result2'] = np.where(df_comparation.iloc[:, 1] == df_comparation.iloc[:, 4], 'OK', 'ERROR')\n",
    "\n",
    "# We save here the rows where errors in gene symbol have been found\n",
    "    gene_error = df_comparation[df_comparation['result1'].str.match('ERROR')]\n",
    "    gene_error = gene_error.drop(gene_error.columns[[1, 4, 5, 6]], axis=1)\n",
    "\n",
    "# We save here the rows where errors in location have been found\n",
    "    location_error = df_comparation[df_comparation['result2'].str.match('ERROR')]\n",
    "    location_error = location_error.drop(location_error.columns[[0, 3, 5, 6]], axis=1)\n",
    "    return gene_error, location_error\n",
    "\n",
    "def gene_loc_error(html_file, location_error, gene_error):\n",
    "    variant2locations = location_error.set_index('ID').T.to_dict('list')\n",
    "    # we save this information in a dict with the variant where the error has been found as key\n",
    "    # Plus in a list the gene simbol of the input and the gene simbol of the API as value.\n",
    "    gene2gene_API = gene_error.set_index('ID').T.to_dict('list')\n",
    "\n",
    "    # If there is not any mismatch found, one document will be created saying\n",
    "    No_error_found = 'Congratulation, no errors have been found in your CSV with regard to gene_symbol and location\\n'\n",
    "    # If there are mismathes found, one document will be created saying\n",
    "    Errors_found = 'If you are reading this message, then the CSVtoVCF application detected some error(s) or mismatch(s) in the Gene Symbol and/or the location of your variants as compared to data retrieved from API Emsembl (VEP).\\n Below the differences found are shown.\\n\\n'\n",
    "    # Plus the mismatch\n",
    "\n",
    "    # Number of locations mismatches\n",
    "    how_many_location_error = 'Number of location mismatch(s) found:', str(len(location_error))\n",
    "    how_many_location_error = ' '.join(how_many_location_error)\n",
    "\n",
    "\n",
    "    # Now, show these mismatchs in this order\n",
    "    header_location = '<th class=\"col_heading level0 col0\" >Your location</th> <th class=\"col_heading level0 col1\" >Variant</th>  <th class=\"col_heading level0 col2\" >API location</th>'\n",
    "    html = html_file\n",
    "\n",
    "# Now with these dicts we can write the inform_error.txt\n",
    "# If both dicts are empty\n",
    "# then create the txt saying there is not error found\n",
    "# If one or both dict contain data\n",
    "# that means mismatch found in the input\n",
    "\n",
    "# Now the same with gene_symbol\n",
    "\n",
    "# Number of gene_symbol mismatchs\n",
    "    how_many_GS_error = 'Number of Gene Symbol mismatch(s) found:', str(len(gene_error))\n",
    "    how_many_GS_error = ' '.join(how_many_GS_error)\n",
    "\n",
    "# Now, show these mismatchs in this order\n",
    "\n",
    "    header_gene_symbol = '<th class=\"col_heading level0 col0\" >Your Gene_symbol</th> <th class=\"col_heading level0 col1\" >Variant</th>  <th class=\"col_heading level0 col2\" >API Gene_symbol</th>'\n",
    "\n",
    "\n",
    "    location_title = \"#####LOCATION'S MISMATCH(S) ###### \\n\\n\"\n",
    "    gene_title = \"\\n\\n##### GENE_SYMBOL'S MISMATCH(S) #######\\n\\n\"\n",
    "\n",
    "# Here we introduce a condition, if dicts are empty\n",
    "# write No_error_found\n",
    "# if dicts are not empty, write all this information\n",
    "\n",
    "\n",
    "    if (bool(variant2locations) == False) and (bool(gene2gene_API) == False):\n",
    "        html += '<div class=errors_summary><b>\\n'+ No_error_found +'</b></div>\\n'\n",
    "        html += \"</body>\"\n",
    "        html += \"</html>\"\n",
    "    #    with open(output_inform, 'w') as out:\n",
    "    #        out.writelines([No_error_found])\n",
    "    else:\n",
    "        print(\"errors found - refer to output error report\")\n",
    "    # Write location\n",
    "\n",
    "    #with open(output_inform, 'w') as out:\n",
    "    #    out.writelines([Errors_found, location_title, how_many_location_error, separator, header_location])\n",
    "    #location_error.to_csv(output_inform,\n",
    "    #                      header=None, index=None, sep='\\t', mode='a')\n",
    "\n",
    "    # Now, with Gene_symbol\n",
    "\n",
    "    #with open(output_inform, 'a') as out:\n",
    "    #    out.writelines([gene_title, how_many_GS_error, separator, header_gene_symbol])\n",
    "    #gene_error.to_csv(output_inform,\n",
    "    #                  header=None, index=None, sep='\\t', mode='a')\n",
    "        html += '<div class=errors_summary>\\n'\n",
    "        html += '<p>'+Errors_found+'</p>'\n",
    "        html += '<div class=error_sum1>\\n'\n",
    "        html += '<p style=\"text-align:center\">'+ gene_title +'</p>\\n'\n",
    "        html += '<p style=\"text-align:center\">'+ how_many_GS_error +'</p>\\n'\n",
    "        html += '<table class=errors>\\n'+ header_gene_symbol\n",
    "        for j in range(len(gene_error)):\n",
    "            html+= '<tr>\\n'\n",
    "            for col2 in gene_error.columns:\n",
    "                value2 = gene_error.iloc[j][col2]\n",
    "                html +='<td>\\n'+str(value2)+'\\n</td>\\n'\n",
    "            html+= '</tr>\\n'\n",
    "        html +='</table>\\n'\n",
    "        html += '</div>\\n'\n",
    "        html += '<div class=error_sum2>\\n'\n",
    "        html += '<p style=\"text-align:center\">'+ location_title +'</p>\\n'\n",
    "        html += '<p style=\"text-align:center\">'+ how_many_location_error +'</p>\\n'\n",
    "        html += '<table class=errors>\\n'+ header_location\n",
    "        for i in range(len(location_error)):\n",
    "            html+= '<tr>\\n'\n",
    "            for col in location_error.columns:\n",
    "                value = location_error.iloc[i][col]    \n",
    "                html +='<td>\\n'+str(value)+'\\n</td>\\n'\n",
    "            html+= '</tr>\\n'\n",
    "        html +='</table>\\n'\n",
    "        html += '</div>\\n'\n",
    "        html += '</div>\\n'\n",
    "        html += '</body>\\n'\n",
    "        html += \"</html>\"\n",
    "    return html\n",
    "\n",
    "if is_stage2 == 'YES':\n",
    "    pattern = re.compile('(NO)|(^$)')\n",
    "    df_filtered = df_exist[~df_exist.exist.str.contains(pattern)]\n",
    "    gene_error, location_error = get_stage2(df_stage1, df_filtered, genome_build)\n",
    "    html_final = gene_loc_error(html_file, location_error, gene_error)\n",
    "    with open (output_inform, 'w') as out:\n",
    "        out.write(html_final)\n",
    "\n",
    "\n",
    "\n",
    "# get error report as html file html file\n",
    "def finish_html(html_file):\n",
    "    text = \"End of error report. CSVtoVCF can also check if your GENE symbol and chromosome \" \\\n",
    "           \"coordinates match with the GENE and location retrieved by API. \" \\\n",
    "           \"To enable this function make sure that your gene names are located under a column with a title 'GENE'\" \\\n",
    "           \"and chromosome coordinates matches example formal (e.g. 17:656767), the column title for this must be 'COORDINATES'.\"\n",
    "    html_file += '<div class=errors_summary><b>\\n'+ text +'</b></div>\\n'\n",
    "    html_file += \"</body>\"\n",
    "    html_file += \"</html>\"\n",
    "    return html_file\n",
    "if is_stage2 == 'NO':\n",
    "    finished_html = finish_html(html_file)\n",
    "    with open (output_inform, 'w') as out:\n",
    "        out.write(finished_html)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Introduce in name, the name of your new file\n",
    "name_of_VCF = \"your_\" + input_file_name + \"_converted_into_VCF.vcf\"\n",
    "\n",
    "output_inform = actual_path + folder_output + name_of_VCF\n",
    "\n",
    "# This will  create directory + file name\n",
    "# completeName2 = os.path.join(save_path, name_of_file2 + \".vcf\")\n",
    "\n",
    "# Now, let's create the metadata information\n",
    "file_format = '##fileformat=VCFv4.3\\n'\n",
    "today = date.today()\n",
    "d1 = today.strftime(\"%Y-%m-%d\")\n",
    "file_Date = '##fileDate=' + d1 + '\\n'\n",
    "\n",
    "# And now the header of the columns\n",
    "#header_line = '#CHROM\\tPOS\\tID\\tREF\\tALT\\tSTRAND\\tQUAL\\tFILTER\\tINFO\\n'\n",
    "header_line = '#CHROM\\tPOS\\tID\\tREF\\tALT\\tSTRAND\\tGENE\\tINFO\\n'\n",
    "\n",
    "# Now, we write meta-information line and header line\n",
    "with open(output_inform, 'w') as out:\n",
    "    out.writelines([file_format, file_Date, header_line])\n",
    "\n",
    "# Let's modify the df_stage1 to be added to the VCF file\n",
    "df_stage2 = df_stage1\n",
    "df_stage2['INFO'] = '.'\n",
    "\n",
    "def convert(v):\n",
    "    try:\n",
    "        return int(v)\n",
    "    except ValueError:\n",
    "        return v\n",
    "\n",
    "\n",
    "def get_vcf_genome(data_frame, genome_build):\n",
    "    choice = {\"a\": \"POS_HG19\", \"b\": \"POS_HG38\"}\n",
    "    data_frame2 = pd.DataFrame([convert(c) for c in l] for l in data_frame.values).sort_values([1, 2], ascending=(True, True))\n",
    "    data_frame2.columns = data_frame.columns\n",
    "    data_frame2 = data_frame2.sort_values(['CHROM', choice[genome_build]])\n",
    "    #data_frame['CHROM'] = data_frame['CHROM'].astype(float)\n",
    "    #data_frame[choice[genome_build]] = data_frame[choice[genome_build]].astype(float)\n",
    "    #data_frame = data_frame.sort_values([ 'CHROM', choice[genome_build]], ascending=(True, True))\n",
    "    #data_frame = data_frame[['CHROM', choice[genome_build], 'ID', 'REF', 'ALT', 'STRAND', 'QUAL', 'FILTER', 'INFO', 'Gene_API']]\n",
    "    data_frame2 = data_frame2[\n",
    "        ['CHROM', choice[genome_build], 'ID', 'REF', 'ALT', 'STRAND', 'Gene_API', 'INFO']]\n",
    "    data_frame2 = data_frame2.replace(to_replace=\"nan\",\n",
    "                              value=\".\")\n",
    "    return data_frame2\n",
    "\n",
    "df_stage3 = get_vcf_genome(df_stage2, genome_vcf_out)\n",
    "\n",
    "df_stage3.to_csv(output_inform,\n",
    "                 header=None, index=None, sep='\\t', mode='a')\n",
    "\n",
    "print('Ta-daa!')\n",
    "print('Have a look in the output folder')\n",
    "print('Your files are ready!')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# TESTING #\n",
    "\n",
    "\"\"\"\n",
    "To compare our results, the REF values generated in the Emsembl API \n",
    "are compared with another API called MyVariantINFO.\n",
    "\n",
    "This API only work with genome building GH19.\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "if df_stage3.columns[1] == 'POS_HG38':\n",
    "    for index, row in df_stage3.iterrows():\n",
    "        chrom = row['CHROM']\n",
    "        pos = row['POS_HG38']\n",
    "        df_stage3.loc[index,[\"POS_HG19\"]] = conv38to19[chrom][pos][0][1]\n",
    "    df_stage3 = df_stage3[['CHROM', 'POS_HG19', 'ID', 'REF', 'ALT',\"STRAND\",\"Gene_API\", \"INFO\"]]\n",
    "    df_stage3[\"POS_HG19\"] = df_stage3[\"POS_HG19\"].astype(int)\n",
    "\n",
    "\n",
    "\n",
    "print(\"Running the test...\\n\")\n",
    "print(\"This test takes the results generated in the output VCF performs a new request in an different API. \\n For a variant given If some of its results matched in both APIs, we asummed that this variant is correct.\\n\")\n",
    "\n",
    "\"\"\"\n",
    "To compare our results, the REF values generated in the Emsembl API \n",
    "are compared with another API called MyVariantINFO.\n",
    "\"\"\"\n",
    "\n",
    "# To create a new column with the input needed for requesting in the new API\n",
    "for idx, value in df_stage3.iloc[:,2].iteritems():\n",
    "    if \">\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + df_stage3.loc[idx, [\"REF\"]][0] + \">\" + df_stage3.loc[idx, [\"ALT\"]][0]\n",
    "    if \"ins\" in value: \n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + \"_\" + ((len(df_stage3.loc[idx,[\"ID\"]].apply(lambda x: x.split(\"ins\")[1])[0]) + df_stage3.loc[idx, [\"POS_HG19\"]][0]).astype(str)) + \"ins\" + df_stage3.loc[8,[\"ID\"]].apply(lambda x: x.split(\"ins\")[1])[0]\n",
    "    if \"del\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = \"chr\"+ df_stage3.loc[idx, [\"CHROM\"]][0].astype(str)+ \":g.\" + df_stage3.loc[idx, [\"POS_HG19\"]][0].astype(str) + \"_\" + ((len(df_stage3.loc[idx,[\"ID\"]].apply(lambda x: x.split(\"del\"))[0][1]) + df_stage3.loc[idx, [\"POS_HG19\"]][0]).astype(str)) + \"del\" \n",
    "    if \"rs\" in value:\n",
    "        df_stage3.loc[idx, [\"TEST\"]] = df_stage3.loc[idx, [\"ID\"]][0]\n",
    "        \n",
    "\n",
    "# Requesting\n",
    "server_rs = \"http://myvariant.info/v1/query?q=\"\n",
    "server = 'http://myvariant.info/v1/variant/'\n",
    "\n",
    "unknow_variants = list()\n",
    "\n",
    "for idx, value in df_stage3.iloc[:,8].iteritems():\n",
    "    if ':g.' in value and \">\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "   \n",
    "    if \"rs\" in value:\n",
    "        r = requests.get(server_rs+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['hits'][0]['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['hits'][0]['vcf']['alt']\n",
    "            \n",
    "    if ':g.' in value and \"del\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "\n",
    "    if ':g.' in value and \"ins\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "\n",
    "# Print results in the terminal\n",
    "\n",
    "\"\"\"\n",
    "Comparint this results, we have noticed that\n",
    "no all variants are in both APIs\n",
    "we inform here how manny variants are not found in both\n",
    "and then of the variants found in both APIs, \n",
    "how many variants match and how many dont match\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "# Test results\n",
    "\n",
    "print(\"From the total number of \",len(df_stage3.index), \"variants counted in your VCF file\", len(unknow_variants), \"are been not found in the MyVariantINFO API\")\n",
    "#df_stage3 = df_stage3.dropna(subset=[\"REF_VariantINFO\"])\n",
    "if not unknow_variants:\n",
    "    pass\n",
    "else:\n",
    "    print(\"Variants do not found in MyVariantINFO API are: \", unknow_variants)\n",
    "df_stage3[\"test_comparation\"] = np.where(df_stage3[\"REF\"]==df_stage3[\"REF_VariantINFO\"],\"OK\",\"ERROR\")\n",
    "\n",
    "ok= df_stage3[df_stage3.test_comparation == \"OK\"].shape[0]\n",
    "print(\"For the rest the variants found in both APIs the \",(len(df_stage3)/ok)*100,\"% of them match in both APIs.\")\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'vcf'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-78-7e5edf8bca4b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mdecoded\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'vcf'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'ref'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m: 'vcf'"
     ]
    }
   ],
   "source": [
    "decoded['vcf']['ref']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG19</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "      <th>TEST</th>\n",
       "      <th>REF_VariantINFO</th>\n",
       "      <th>ALT_VariantINFO</th>\n",
       "      <th>test_comparation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>100951675.0</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "      <td>chr3:g.100951675.0G&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31239170.0</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.31239170.0_31239171.0insT</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35427531.0</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "      <td>chr6:g.35427531.0T&gt;C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555.0</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555.0_117199558.0del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555.0</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>chr7:g.117199555.0_117199558.0del</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>136131429.0</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23896042.0</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "      <td>chr14:g.23896042.0C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48892422.0</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr15:g.48892422.0C&gt;T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>41219633.0</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>chr17:g.41219633.0A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>41223094.0</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM     POS_HG19                                 ID   REF ALT STRAND  \\\n",
       "0      3  100951675.0              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31239170.0    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35427531.0          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117199555.0  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117199555.0  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  136131429.0                         rs56116432     C   T      +   \n",
       "3     14   23896042.0              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48892422.0               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   41219633.0              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   41223094.0                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API INFO                               TEST REF_VariantINFO  \\\n",
       "0    IMPG2    .              chr3:g.100951675.0G>C             NaN   \n",
       "8    HLA-C    .   chr6:g.31239170.0_31239171.0insT             NaN   \n",
       "9    FANCE    .               chr6:g.35427531.0T>C             NaN   \n",
       "6     CFTR    .  chr7:g.117199555.0_117199558.0del             NaN   \n",
       "7     CFTR    .  chr7:g.117199555.0_117199558.0del             NaN   \n",
       "4      ABO    .                         rs56116432               C   \n",
       "3     MYH7    .              chr14:g.23896042.0C>T             NaN   \n",
       "2     FBN1    .              chr15:g.48892422.0C>T             NaN   \n",
       "1    BRCA1    .              chr17:g.41219633.0A>G             NaN   \n",
       "5    BRCA1    .                          rs1799966               T   \n",
       "\n",
       "  ALT_VariantINFO test_comparation  \n",
       "0             NaN            ERROR  \n",
       "8             NaN            ERROR  \n",
       "9             NaN            ERROR  \n",
       "6             NaN            ERROR  \n",
       "7             NaN            ERROR  \n",
       "4               T               OK  \n",
       "3             NaN            ERROR  \n",
       "2             NaN            ERROR  \n",
       "1             NaN            ERROR  \n",
       "5               A               OK  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stage3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG38</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "      <th>POS_HG19</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "      <td>100951675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31271393</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>.</td>\n",
       "      <td>31239170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35459754</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "      <td>35427531</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117559501</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>117199555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117559501</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "      <td>117199555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "      <td>136131429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "      <td>23896042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "      <td>48892422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>41219633</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "      <td>41223094</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG38                                 ID   REF ALT STRAND  \\\n",
       "0      3  101232831              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31271393    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35459754          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117559501  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117559501  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  133256042                         rs56116432     C   T      +   \n",
       "3     14   23426833              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48600225               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   43067616              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   43071077                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API INFO   POS_HG19  \n",
       "0    IMPG2    .  100951675  \n",
       "8    HLA-C    .   31239170  \n",
       "9    FANCE    .   35427531  \n",
       "6     CFTR    .  117199555  \n",
       "7     CFTR    .  117199555  \n",
       "4      ABO    .  136131429  \n",
       "3     MYH7    .   23896042  \n",
       "2     FBN1    .   48892422  \n",
       "1    BRCA1    .   41219633  \n",
       "5    BRCA1    .   41223094  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for idx, value in df_stage3.iloc[:,8].iteritems():\n",
    "    if ':g.' in value and \">\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "   \n",
    "    if \"rs\" in value:\n",
    "        r = requests.get(server_rs+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['hits'][0]['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['hits'][0]['vcf']['alt']\n",
    "            \n",
    "    if ':g.' in value and \"del\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n",
    "\n",
    "    if ':g.' in value and \"ins\" in value:\n",
    "        r = requests.get(server+ (df_stage3.loc[idx, [\"TEST\"]][0]))\n",
    "        if r.status_code == 404:\n",
    "            unknow_variants.append(value)\n",
    "        if r.status_code == 200:\n",
    "            decoded =  r.json()\n",
    "            df_stage3.loc[idx, [\"REF_VariantINFO\"]] = decoded['vcf']['ref']\n",
    "            df_stage3.loc[idx, [\"ALT_VariantINFO\"]] = decoded['vcf']['alt']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG19</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>100951675</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31239170</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35427531</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117199555</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>136131429</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23896042</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48892422</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>41219633</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>41223094</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG19                                 ID   REF ALT STRAND  \\\n",
       "0      3  100951675              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31239170    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35427531          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117199555  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  136131429                         rs56116432     C   T      +   \n",
       "3     14   23896042              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48892422               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   41219633              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   41223094                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API  \n",
       "0    IMPG2  \n",
       "8    HLA-C  \n",
       "9    FANCE  \n",
       "6     CFTR  \n",
       "7     CFTR  \n",
       "4      ABO  \n",
       "3     MYH7  \n",
       "2     FBN1  \n",
       "1    BRCA1  \n",
       "5    BRCA1  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if df_stage3.columns[1] == 'POS_HG38':\n",
    "    for index, row in df_stage3.iterrows():\n",
    "        chrom = row['CHROM']\n",
    "        pos = row['POS_HG38']\n",
    "        df_stage3.loc[index,[\"POS_HG19\"]] = conv38to19[chrom][pos][0][1]\n",
    "    df_stage3 = df_stage3[['CHROM', 'POS_HG19', 'ID', 'REF', 'ALT',\"STRAND\",\"Gene_API\"]]\n",
    "        \n",
    "df_stage3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS_HG38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>48600225</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   POS_HG38\n",
       "2  48600225"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stage3.loc[[2],[\"POS_HG38\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "=========>"
     ]
    }
   ],
   "source": [
    "\n",
    "import time\n",
    "for i in range(10):\n",
    "    print('\\r{}>'.format('='*i), end='')\n",
    "    time.sleep(0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "ppolla =list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG38</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31271393</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35459754</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117559501</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117559501</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG38                                 ID   REF ALT STRAND  \\\n",
       "0      3  101232831              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31271393    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35459754          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117559501  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117559501  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  133256042                         rs56116432     C   T      +   \n",
       "3     14   23426833              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48600225               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   43067616              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   43071077                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API INFO  \n",
       "0    IMPG2    .  \n",
       "8    HLA-C    .  \n",
       "9    FANCE    .  \n",
       "6     CFTR    .  \n",
       "7     CFTR    .  \n",
       "4      ABO    .  \n",
       "3     MYH7    .  \n",
       "2     FBN1    .  \n",
       "1    BRCA1    .  \n",
       "5    BRCA1    .  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_hg19(df_stage3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS_HG38</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>STRAND</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>INFO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31271393</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>-</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>35459754</td>\n",
       "      <td>ENST00000229769:c.1310T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>FANCE</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117559501</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117559501</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CTTC</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>ABO</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM   POS_HG38                                 ID   REF ALT STRAND  \\\n",
       "0      3  101232831              NM_016247.3:c.3183C>G     G   C      -   \n",
       "8      6   31271393    NM_002117.5:c.344-46_344-45insT     C  CA      -   \n",
       "9      6   35459754          ENST00000229769:c.1310T>C     T   C      +   \n",
       "6      7  117559501  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "7      7  117559501  ENST00000003084:c.1431_1433delTTC  CTTC   C      +   \n",
       "4      9  133256042                         rs56116432     C   T      +   \n",
       "3     14   23426833              NM_000257.3:c.1988G>A     C   T      -   \n",
       "2     15   48600225               NM_000138.4:c.356G>A     C   T      -   \n",
       "1     17   43067616              NM_007294.3:c.5066T>C     A   G      -   \n",
       "5     17   43071077                          rs1799966     T   A      +   \n",
       "\n",
       "  Gene_API INFO  \n",
       "0    IMPG2    .  \n",
       "8    HLA-C    .  \n",
       "9    FANCE    .  \n",
       "6     CFTR    .  \n",
       "7     CFTR    .  \n",
       "4      ABO    .  \n",
       "3     MYH7    .  \n",
       "2     FBN1    .  \n",
       "1    BRCA1    .  \n",
       "5    BRCA1    .  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stage3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
