{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stage 1\n",
    "****"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "##################################################################\n",
    "\n",
    "# Welcome to the script CSV2VCF-converter                        #\n",
    "# A assessment of the module of Introduction to Programming      #\n",
    "# By Igor Malashchuk and Manuel Dominguez                        #\n",
    "\n",
    "##################################################################\n",
    "\n",
    "# This programme has been split in a few stages we explain here.\n",
    "# The code is full of comments in which we explain line by line\n",
    "# what we are doing, however we explain a general explanation\n",
    "# plus details stage by stage\n",
    "\n",
    "\n",
    "\n",
    "# This script converts variants stored in a Comma Separated Values (CSV) file\n",
    "# into Variant Call Format (VCF)\n",
    "\n",
    "# STAGE 1\n",
    "\n",
    "## I) Identify 3 (if time maybe more) different ID Variants from a CSV file\n",
    "\n",
    "### Owing to different clinical scientists use a slightly different CSV\n",
    "### We have tried to create a code as flexible as possible\n",
    "### but a we have established minimum requirements we ask here.\n",
    "### The gene symbol and the location of your variant\n",
    "### must be the first and second column respectively of your CSV file.\n",
    "### This is easy to do using Microsoft Excel.\n",
    "### Our code recognise 3 different Variant ID:\n",
    "\n",
    "### AGT:c.803T>C\n",
    "### rs17289390\n",
    "### ENST00000003084:c.1431_1433delTTC (We need to check this)\n",
    "\n",
    "\n",
    "### We have test our script and this support Single Nucleotide Variants, \n",
    "### insertion variants and deletion Variants. \n",
    "\n",
    "### If your CSV have more than one of the mentioned variant ID in the same row\n",
    "### The script will select the first found, starting from the left\n",
    "### After the first, following variants ID will be ignore. \n",
    "\n",
    "### In Stage 1 we have consider the possibility of getting errors in your variant ID.\n",
    "### If this happen, the API requested will provide an error\n",
    "### We have solved this issue with one exception\n",
    "\n",
    "\n",
    "# STAGE 2\n",
    "\n",
    "### We have noticed that the CSV used for years are not uptodate\n",
    "### We will compare for you if your gene symbol and location are correct\n",
    "### This script compares your gene symbol and variant with the equivalent provided by the API.\n",
    "### And creates a txt file informing you error or mismatchs found.\n",
    "### The VCF generate will used directly the gene symbol and the location from the API\n",
    "\n",
    "# STAGE 3 \n",
    "\n",
    "### This stage creates the VCF file according to documentation (VCFv4.3 format)\n",
    "### https://samtools.github.io/hts-specs/VCFv4.2.pdf\n",
    "\n",
    "\n",
    "# STAGE 4\n",
    "\n",
    "### because this is a collaborative project between two people\n",
    "### we have used Jupyter Notebooks and Github.\n",
    "### This stage consists of creating a script that performs the same function using the terminal.\n",
    "\n",
    "\n",
    "# Stage 5\n",
    "\n",
    "### The last stage consists of creating a interface that perfoms the same execution that stage 4.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's import all modules used in this script:\n",
    "import sys\n",
    "import os\n",
    "import requests\n",
    "import chardet\n",
    "import json\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os.path\n",
    "from datetime import date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "your_path = os.getcwd()\n",
    "folder = '/input'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['NM_.csv']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file_list = os.listdir(your_path+folder)\n",
    "file_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GRCh38 coordinates</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Quality</th>\n",
       "      <th>Filter</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>77:43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>14/85</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117199558</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene GRCh38 coordinates                         Transcript Quality  \\\n",
       "0   IMPG2        3:101232831              NM_016247.3:c.3183C>G   34/64   \n",
       "1   BRCA1        17:43067616              NM_007294.3:c.5066T>C  49/131   \n",
       "2    FBN1        15:48600225               NM_000138.4:c.356G>A   46/99   \n",
       "3    MYH7        14:23426833              NM_000257.3:c.1988G>A   8/102   \n",
       "4   BORIS        9:133256042                         rs56116432   1/200   \n",
       "5   BRCA1        77:43071077                          rs1799966   14/85   \n",
       "6    CFTR        7:117199558  ENST00000003084:c.1431_1433delTTC  99/100   \n",
       "7    CFTR        7:117559502  ENST00000003084:c.1431_1433delTTC  99/100   \n",
       "8   HLA-C         6:31271394    NM_002117.5:c.344-46_344-45insT   34/64   \n",
       "9   HLA-C         6:31271394                          rs9281300   34/64   \n",
       "10    LPL         8:19956018               NM_000237.2:c.953A>G   34/64   \n",
       "11   TSC2         16:2088197                          rs1799758   34/64   \n",
       "\n",
       "     Filter  \n",
       "0      PASS  \n",
       "1      PASS  \n",
       "2      PASS  \n",
       "3   NO PASS  \n",
       "4   NO PASS  \n",
       "5      PASS  \n",
       "6      PASS  \n",
       "7      PASS  \n",
       "8      PASS  \n",
       "9      PASS  \n",
       "10     PASS  \n",
       "11     PASS  "
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To read input\n",
    "\n",
    "# This code is able to recognise the 3 type of ID variant mention above contained in a CSV file.\n",
    "\n",
    "CSV_input = '/Users/manolodominguez/Desktop/git-repos/STP_mini_projects/Igor-Manuel/Inputs/NM_.csv'\n",
    "\n",
    "# input to pd data frame\n",
    "\n",
    "# This is to avoid UnicodeDecodeError: 'utf-8'\n",
    "with open(CSV_input, 'rb') as f:\n",
    "    result = chardet.detect(f.read())\n",
    "\n",
    "# This not work 100% of the times. To avoid UnicodeDecodeError: 'utf-8'\n",
    "# We strongly recommend that the input be .csv\n",
    "\n",
    "df = pd.read_csv(CSV_input, encoding=result['encoding'])\n",
    "\n",
    "# We will need a extra copy to avoid repetitions\n",
    "df_draft = df\n",
    "\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Here we read and found what we want from the CSV\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Is the DataFrame empty?  True\n"
     ]
    }
   ],
   "source": [
    "##### Reading CSV file values and looking for variants IDs ######\n",
    "\n",
    "# Our programme recognise hgvs_notation (e.g AGT:c.803T>C and ENST00000003084:c.1431_1433delTTC) \n",
    "# and also dbSNP ID (e.g rs17289390)\n",
    "\n",
    "\n",
    "# Find Variant Transcript ':c.' in CSV\n",
    "Transcript = df_draft[df_draft.apply(lambda x:x.str.contains(\":c.\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "\n",
    "# Now, we save the results found in a dict key=index and value=variand ID\n",
    "if Transcript.empty == False:\n",
    "    ind = Transcript.index.to_list()\n",
    "    vals = list(Transcript.stack().values)\n",
    "    row2Transcript = dict(zip(ind, vals))\n",
    "    # We need to remove the row where rs has been found to avoid repetitions\n",
    "    # In case in same row more than one kind of ID Variant is stored ('e.g :c.' and '(:p.)')   \n",
    "    for index, Transcript  in row2Transcript.items(): \n",
    "        # This will be done in df_draft\n",
    "        df_draft = df_draft.drop(index)\n",
    "        \n",
    "# Same but now with dbSNP variant ID\n",
    "rs = df[df.apply(lambda x:x.str.contains(\"rs\\d+\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "\n",
    "#rs = df_draft[df_draft.apply(lambda x:x.str.contains(\"rs\"))].dropna(how='all').dropna(axis=1, how='all')\n",
    "\n",
    "# Now, we save the results found in a dict key=index and value=variand ID\n",
    "if rs.empty == False:\n",
    "    ind = rs.index.to_list()\n",
    "    vals = list(rs.stack().values)\n",
    "    row2rs = dict(zip(ind, vals))\n",
    "    # We need to remove the row where rs has been found to avoid repetitions\n",
    "    # In case in same row more than one kind of ID Variant is stored ('e.g :c.' and '(:p.)')   \n",
    "    for index, rs  in row2rs.items(): \n",
    "        # This will be done in df_draft\n",
    "        df_draft = df_draft.drop(index)\n",
    "\n",
    "\n",
    "        \n",
    "# df_draft empty means every row contained a rs or :c. ID\n",
    "print('Is the DataFrame empty? ', df_draft.empty)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Let's see if last code worked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{4: 'rs56116432', 5: 'rs1799966', 9: 'rs9281300', 11: 'rs1799758'}"
      ]
     },
     "execution_count": 202,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2rs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'NM_016247.3:c.3183C>G',\n",
       " 1: 'NM_007294.3:c.5066T>C',\n",
       " 2: 'NM_000138.4:c.356G>A',\n",
       " 3: 'NM_000257.3:c.1988G>A',\n",
       " 6: 'ENST00000003084:c.1431_1433delTTC',\n",
       " 7: 'ENST00000003084:c.1431_1433delTTC',\n",
       " 8: 'NM_002117.5:c.344-46_344-45insT',\n",
       " 10: 'NM_000237.2:c.953A>G'}"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2Transcript"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Now, the request API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {},
   "outputs": [],
   "source": [
    "row2gene_symbol = dict()\n",
    "row2CHROM = dict()\n",
    "row2POS = dict()\n",
    "row2ALT = dict()\n",
    "row2REF = dict()\n",
    "\n",
    "\n",
    "server = 'https://rest.ensembl.org/vep/human/hgvs/'\n",
    "\n",
    "for row, transcript in row2Transcript.items():\n",
    "\n",
    "    r = requests.get(server+transcript, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "    decoded = r.json()[0]\n",
    "    \n",
    "# For row2gene_symbol\n",
    "    row2gene_symbol[row]= decoded[\"transcript_consequences\"][0]['gene_symbol']\n",
    "\n",
    "# For row2CHROM\n",
    "    row2CHROM[row] = decoded[\"seq_region_name\"]\n",
    "    \n",
    "# For row2POS \n",
    "    row2POS[row] = decoded[\"start\"]\n",
    "    \n",
    "# For row2ALT \n",
    "    row2ALT[row] = decoded['allele_string'].split('/')[1]\n",
    "# This also work: decoded[\"transcript_consequences\"][0]['variant_allele']\n",
    "\n",
    "# For row2REF\n",
    "    row2REF[row] = decoded['allele_string'].split('/')[0]\n",
    "    \n",
    "########################################################################    \n",
    "# Here is where we need to put some exception if variant ID is unknown #\n",
    "########################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Let's see if the ditc work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'IMPG2',\n",
       " 1: 'BRCA1',\n",
       " 2: 'FBN1',\n",
       " 3: 'MYH7',\n",
       " 6: 'CFTR',\n",
       " 7: 'CFTR',\n",
       " 8: 'HLA-C',\n",
       " 10: 'LPL'}"
      ]
     },
     "execution_count": 319,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2gene_symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: '3', 1: '17', 2: '15', 3: '14', 6: '7', 7: '7', 8: '6', 10: '8'}"
      ]
     },
     "execution_count": 320,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2CHROM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 101232831,\n",
       " 1: 43067616,\n",
       " 2: 48600225,\n",
       " 3: 23426833,\n",
       " 6: 117559502,\n",
       " 7: 117559502,\n",
       " 8: 31271394,\n",
       " 10: 19956018}"
      ]
     },
     "execution_count": 321,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2POS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'G', 1: 'C', 2: 'A', 3: 'A', 6: '-', 7: '-', 8: 'T', 10: 'G'}"
      ]
     },
     "execution_count": 322,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2ALT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'C', 1: 'T', 2: 'G', 3: 'G', 6: 'TTC', 7: 'TTC', 8: '-', 10: 'A'}"
      ]
     },
     "execution_count": 323,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2REF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, we need to do the same with row2rs\n",
    "\n",
    "server = \"https://rest.ensembl.org/vep/human/id/\"\n",
    "\n",
    "interrogation = '?'\n",
    "\n",
    "for row, rs in row2rs.items():\n",
    "\n",
    "    r = requests.get(server+rs+interrogation, headers={ \"Content-Type\" : \"application/json\"})\n",
    "    if not r.ok:\n",
    "        r.raise_for_status()\n",
    "        sys.exit()\n",
    "    decoded = r.json()[0]\n",
    "    \n",
    "    \n",
    "# For row2gene_symbol\n",
    "    row2gene_symbol[row]= decoded['transcript_consequences'][0]['gene_symbol']\n",
    "\n",
    "# For row2CHROM\n",
    "    row2CHROM[row] = decoded['seq_region_name']\n",
    "    \n",
    "# For row2POS \n",
    "    row2POS[row] = decoded['start']\n",
    "    \n",
    "# For row2ALT \n",
    "    row2ALT[row] = decoded['allele_string'].split('/')[1]\n",
    "\n",
    "# For row2REF\n",
    "    row2REF[row] = decoded['allele_string'].split('/')[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'IMPG2',\n",
       " 1: 'BRCA1',\n",
       " 2: 'FBN1',\n",
       " 3: 'MYH7',\n",
       " 6: 'CFTR',\n",
       " 7: 'CFTR',\n",
       " 8: 'HLA-C',\n",
       " 10: 'LPL',\n",
       " 4: 'ABO',\n",
       " 5: 'BRCA1',\n",
       " 9: 'HLA-C',\n",
       " 11: 'TSC2'}"
      ]
     },
     "execution_count": 325,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2gene_symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 326,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: '3',\n",
       " 1: '17',\n",
       " 2: '15',\n",
       " 3: '14',\n",
       " 6: '7',\n",
       " 7: '7',\n",
       " 8: '6',\n",
       " 10: '8',\n",
       " 4: '9',\n",
       " 5: '17',\n",
       " 9: '6',\n",
       " 11: '16'}"
      ]
     },
     "execution_count": 326,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2CHROM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 327,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 101232831,\n",
       " 1: 43067616,\n",
       " 2: 48600225,\n",
       " 3: 23426833,\n",
       " 6: 117559502,\n",
       " 7: 117559502,\n",
       " 8: 31271394,\n",
       " 10: 19956018,\n",
       " 4: 133256042,\n",
       " 5: 43071077,\n",
       " 9: 31271394,\n",
       " 11: 2088197}"
      ]
     },
     "execution_count": 327,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2POS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 328,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'G',\n",
       " 1: 'C',\n",
       " 2: 'A',\n",
       " 3: 'A',\n",
       " 6: '-',\n",
       " 7: '-',\n",
       " 8: 'T',\n",
       " 10: 'G',\n",
       " 4: 'T',\n",
       " 5: 'A',\n",
       " 9: 'A',\n",
       " 11: 'AG'}"
      ]
     },
     "execution_count": 328,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2ALT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 329,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'C',\n",
       " 1: 'T',\n",
       " 2: 'G',\n",
       " 3: 'G',\n",
       " 6: 'TTC',\n",
       " 7: 'TTC',\n",
       " 8: '-',\n",
       " 10: 'A',\n",
       " 4: 'C',\n",
       " 5: 'T',\n",
       " 9: '-',\n",
       " 11: 'AGTGAG'}"
      ]
     },
     "execution_count": 329,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row2REF"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This looks ok"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 330,
   "metadata": {},
   "outputs": [],
   "source": [
    "# To finish this stage, \n",
    "# Let's save our dicts in a df\n",
    "\n",
    "column_names = [\"Gene_API\", 'CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO']\n",
    "df_stage1 = pd.DataFrame(columns = column_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 331,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABO</td>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>14/85</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7</td>\n",
       "      <td>117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>-</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7</td>\n",
       "      <td>117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>-</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16</td>\n",
       "      <td>2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>AGTGAG</td>\n",
       "      <td>AG</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene_API CHROM        POS                                 ID     REF ALT  \\\n",
       "0     IMPG2     3  101232831              NM_016247.3:c.3183C>G       C   G   \n",
       "1     BRCA1    17   43067616              NM_007294.3:c.5066T>C       T   C   \n",
       "2      FBN1    15   48600225               NM_000138.4:c.356G>A       G   A   \n",
       "3      MYH7    14   23426833              NM_000257.3:c.1988G>A       G   A   \n",
       "4       ABO     9  133256042                         rs56116432       C   T   \n",
       "5     BRCA1    17   43071077                          rs1799966       T   A   \n",
       "6      CFTR     7  117559502  ENST00000003084:c.1431_1433delTTC     TTC   -   \n",
       "7      CFTR     7  117559502  ENST00000003084:c.1431_1433delTTC     TTC   -   \n",
       "8     HLA-C     6   31271394    NM_002117.5:c.344-46_344-45insT       -   T   \n",
       "9     HLA-C     6   31271394                          rs9281300       -   A   \n",
       "10      LPL     8   19956018               NM_000237.2:c.953A>G       A   G   \n",
       "11     TSC2    16    2088197                          rs1799758  AGTGAG  AG   \n",
       "\n",
       "      QUAL   FILTER  \n",
       "0    34/64     PASS  \n",
       "1   49/131     PASS  \n",
       "2    46/99     PASS  \n",
       "3    8/102  NO PASS  \n",
       "4    1/200  NO PASS  \n",
       "5    14/85     PASS  \n",
       "6   99/100     PASS  \n",
       "7   99/100     PASS  \n",
       "8    34/64     PASS  \n",
       "9    34/64     PASS  \n",
       "10   34/64     PASS  \n",
       "11   34/64     PASS  "
      ]
     },
     "execution_count": 331,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_stage1 = pd.DataFrame.from_dict(dict(zip(column_names, [row2gene_symbol, row2CHROM, row2POS,df[\"Transcript\"],row2REF ,row2ALT, df[\"Quality\"], df[\"Filter\"]  ])))\n",
    "df_stage1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# End of stage 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stage 2\n",
    "****"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In Stage 2 we compare Gene and location between input and API Emsembl.\n",
    "# And we create a txt file to provide the errors/mismachts found.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, let's prepare where our information file is going to be create.\n",
    "\n",
    "# Introce in path the directory where we wish to create the new CVF\n",
    "save_path = '/Users/manolodominguez/Desktop/git-repos/STP_mini_projects/Igor-Manuel/Outputs'\n",
    "\n",
    "# Introduce in name, the name of your new file\n",
    "name_of_file= \"inform_error\"\n",
    "\n",
    "# This will  create directory + file name\n",
    "completeName = os.path.join(save_path, name_of_file+\".txt\")         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 334,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3:101232831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "      <td>17:43067616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>15:48600225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>14:23426833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ABO</td>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>9:133256042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17</td>\n",
       "      <td>43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>14/85</td>\n",
       "      <td>PASS</td>\n",
       "      <td>17:43071077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7</td>\n",
       "      <td>117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>-</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>7:117559502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7</td>\n",
       "      <td>117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>-</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>7:117559502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6:31271394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>6:31271394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8</td>\n",
       "      <td>19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>8:19956018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16</td>\n",
       "      <td>2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>AGTGAG</td>\n",
       "      <td>AG</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>16:2088197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Gene_API CHROM        POS                                 ID     REF ALT  \\\n",
       "0     IMPG2     3  101232831              NM_016247.3:c.3183C>G       C   G   \n",
       "1     BRCA1    17   43067616              NM_007294.3:c.5066T>C       T   C   \n",
       "2      FBN1    15   48600225               NM_000138.4:c.356G>A       G   A   \n",
       "3      MYH7    14   23426833              NM_000257.3:c.1988G>A       G   A   \n",
       "4       ABO     9  133256042                         rs56116432       C   T   \n",
       "5     BRCA1    17   43071077                          rs1799966       T   A   \n",
       "6      CFTR     7  117559502  ENST00000003084:c.1431_1433delTTC     TTC   -   \n",
       "7      CFTR     7  117559502  ENST00000003084:c.1431_1433delTTC     TTC   -   \n",
       "8     HLA-C     6   31271394    NM_002117.5:c.344-46_344-45insT       -   T   \n",
       "9     HLA-C     6   31271394                          rs9281300       -   A   \n",
       "10      LPL     8   19956018               NM_000237.2:c.953A>G       A   G   \n",
       "11     TSC2    16    2088197                          rs1799758  AGTGAG  AG   \n",
       "\n",
       "      QUAL   FILTER     Location  \n",
       "0    34/64     PASS  3:101232831  \n",
       "1   49/131     PASS  17:43067616  \n",
       "2    46/99     PASS  15:48600225  \n",
       "3    8/102  NO PASS  14:23426833  \n",
       "4    1/200  NO PASS  9:133256042  \n",
       "5    14/85     PASS  17:43071077  \n",
       "6   99/100     PASS  7:117559502  \n",
       "7   99/100     PASS  7:117559502  \n",
       "8    34/64     PASS   6:31271394  \n",
       "9    34/64     PASS   6:31271394  \n",
       "10   34/64     PASS   8:19956018  \n",
       "11   34/64     PASS   16:2088197  "
      ]
     },
     "execution_count": 334,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# To compare input with API result\n",
    "# We need to create a new column from CHROM + ':' + POS\n",
    "\n",
    "df_stage1 = df_stage1.applymap(str)\n",
    "df_stage1['Location'] = df_stage1['CHROM'].str.cat(df_stage1['POS'], sep=':')\n",
    "df_stage1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, we can compare Gene and location of the input againt Gene and location\n",
    "# from the API\n",
    "\n",
    "# Let's create a new df for comparison purpose\n",
    "\n",
    "df_comparation =  pd.concat([df.iloc[:, 0:3], df_stage1[['Gene_API','Location']]], axis=1)\n",
    "#df_comparation = df_comparation.set_index(\"Transcript\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GRCh38 coordinates</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>ABO</td>\n",
       "      <td>9:133256042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>77:43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43071077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117199558</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene GRCh38 coordinates                         Transcript Gene_API  \\\n",
       "0   IMPG2        3:101232831              NM_016247.3:c.3183C>G    IMPG2   \n",
       "1   BRCA1        17:43067616              NM_007294.3:c.5066T>C    BRCA1   \n",
       "2    FBN1        15:48600225               NM_000138.4:c.356G>A     FBN1   \n",
       "3    MYH7        14:23426833              NM_000257.3:c.1988G>A     MYH7   \n",
       "4   BORIS        9:133256042                         rs56116432      ABO   \n",
       "5   BRCA1        77:43071077                          rs1799966    BRCA1   \n",
       "6    CFTR        7:117199558  ENST00000003084:c.1431_1433delTTC     CFTR   \n",
       "7    CFTR        7:117559502  ENST00000003084:c.1431_1433delTTC     CFTR   \n",
       "8   HLA-C         6:31271394    NM_002117.5:c.344-46_344-45insT    HLA-C   \n",
       "9   HLA-C         6:31271394                          rs9281300    HLA-C   \n",
       "10    LPL         8:19956018               NM_000237.2:c.953A>G      LPL   \n",
       "11   TSC2         16:2088197                          rs1799758     TSC2   \n",
       "\n",
       "       Location  \n",
       "0   3:101232831  \n",
       "1   17:43067616  \n",
       "2   15:48600225  \n",
       "3   14:23426833  \n",
       "4   9:133256042  \n",
       "5   17:43071077  \n",
       "6   7:117559502  \n",
       "7   7:117559502  \n",
       "8    6:31271394  \n",
       "9    6:31271394  \n",
       "10   8:19956018  \n",
       "11   16:2088197  "
      ]
     },
     "execution_count": 336,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "metadata": {},
   "outputs": [],
   "source": [
    "# New two columns as result of the comparation\n",
    "df_comparation['result1'] = np.where(df_comparation.iloc[:, 0] == df_comparation.iloc[:, 3], 'OK', 'ERROR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 338,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_comparation['result2'] = np.where(df_comparation.iloc[:, 1] == df_comparation.iloc[:, 4], 'OK', 'ERROR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 339,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>GRCh38 coordinates</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>Location</th>\n",
       "      <th>result1</th>\n",
       "      <th>result2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>ABO</td>\n",
       "      <td>9:133256042</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BRCA1</td>\n",
       "      <td>77:43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43071077</td>\n",
       "      <td>OK</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117199558</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "      <td>OK</td>\n",
       "      <td>ERROR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "      <td>OK</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     Gene GRCh38 coordinates                         Transcript Gene_API  \\\n",
       "0   IMPG2        3:101232831              NM_016247.3:c.3183C>G    IMPG2   \n",
       "1   BRCA1        17:43067616              NM_007294.3:c.5066T>C    BRCA1   \n",
       "2    FBN1        15:48600225               NM_000138.4:c.356G>A     FBN1   \n",
       "3    MYH7        14:23426833              NM_000257.3:c.1988G>A     MYH7   \n",
       "4   BORIS        9:133256042                         rs56116432      ABO   \n",
       "5   BRCA1        77:43071077                          rs1799966    BRCA1   \n",
       "6    CFTR        7:117199558  ENST00000003084:c.1431_1433delTTC     CFTR   \n",
       "7    CFTR        7:117559502  ENST00000003084:c.1431_1433delTTC     CFTR   \n",
       "8   HLA-C         6:31271394    NM_002117.5:c.344-46_344-45insT    HLA-C   \n",
       "9   HLA-C         6:31271394                          rs9281300    HLA-C   \n",
       "10    LPL         8:19956018               NM_000237.2:c.953A>G      LPL   \n",
       "11   TSC2         16:2088197                          rs1799758     TSC2   \n",
       "\n",
       "       Location result1 result2  \n",
       "0   3:101232831      OK      OK  \n",
       "1   17:43067616      OK      OK  \n",
       "2   15:48600225      OK      OK  \n",
       "3   14:23426833      OK      OK  \n",
       "4   9:133256042   ERROR      OK  \n",
       "5   17:43071077      OK   ERROR  \n",
       "6   7:117559502      OK   ERROR  \n",
       "7   7:117559502      OK      OK  \n",
       "8    6:31271394      OK      OK  \n",
       "9    6:31271394      OK      OK  \n",
       "10   8:19956018      OK      OK  \n",
       "11   16:2088197      OK      OK  "
      ]
     },
     "execution_count": 339,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_comparation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 340,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We save here the rows where errors have been in gene\n",
    "gene_error = df_comparation[df_comparation['result1'].str.match('ERROR')]\n",
    "gene_error = gene_error.drop(gene_error.columns[[1,4,5,6]], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Gene_API</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BORIS</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>ABO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene  Transcript Gene_API\n",
       "4  BORIS  rs56116432      ABO"
      ]
     },
     "execution_count": 341,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rs56116432': ['BORIS', 'ABO']}"
      ]
     },
     "execution_count": 342,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# we save this information in a dict with the variant where the error has been found as key\n",
    "# Plus in a list the gene simbol of the input and the gene simbol of the API as value.\n",
    "gene2gene_API = gene_error.set_index('Transcript').T.to_dict('list')\n",
    "gene2gene_API"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Same operation for location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GRCh38 coordinates</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>77:43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>17:43071077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7:117199558</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>7:117559502</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  GRCh38 coordinates                         Transcript     Location\n",
       "5        77:43071077                          rs1799966  17:43071077\n",
       "6        7:117199558  ENST00000003084:c.1431_1433delTTC  7:117559502"
      ]
     },
     "execution_count": 344,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "location_error = df_comparation[df_comparation['result2'].str.match('ERROR')]\n",
    "location_error = location_error.drop(location_error.columns[[0,3,5,6]], axis=1)\n",
    "location_error"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 345,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rs1799966': ['77:43071077', '17:43071077'],\n",
       " 'ENST00000003084:c.1431_1433delTTC': ['7:117199558', '7:117559502']}"
      ]
     },
     "execution_count": 345,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variant2locations = location_error.set_index('Transcript').T.to_dict('list')\n",
    "variant2locations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now with dicts we can write the inform_error.txt\n",
    "# If both dicts are empty\n",
    "# then create the txt saying there is not error found\n",
    "# If one or both dict contain data\n",
    "# then says the errors found\n",
    "\n",
    "\n",
    "# If there is not any mismatch found, one document will be created saying \n",
    "No_error_found = 'Congratulation, no errors have been found in your CSV with regard to gene symbol and location\\n'\n",
    "\n",
    "\n",
    "# If there are mismathes found, one document will be created saying\n",
    "Errors_found = 'If you are reading this document is because some error(s) or mismatch(s) has been found between the Gene Symbol and/or the location of your variants\\nWhen these has been compared with API Emsembl.\\nWe show you the differences found in this document.\\n\\n'\n",
    "# Plus the mismatch\n",
    "\n",
    "\n",
    "# Number of locations mismatches \n",
    "how_many_location_error = 'Number of location mismachts found:', str(len(location_error))\n",
    "how_many_location_error = ' '.join(how_many_location_error)\n",
    "\n",
    "# Now, show these mismatchs in this order\n",
    "\n",
    "header_location = 'Your location     variant    API location\\n'\n",
    "\n",
    "# Now the same with gene_symbol\n",
    "\n",
    "# Number of gene_symbol mismatchs\n",
    "how_many_GS_error = 'Number of Gene Symbol mismachts found:', str(len(gene_error))\n",
    "how_many_GS_error = ' '.join(how_many_GS_error)\n",
    "\n",
    "# Now, show these mismatchs in this order\n",
    "\n",
    "header_gene_symbol = 'Your Gene_symbol,     Variant,       API Gene_symbol\\n'\n",
    "separator = '\\n'\n",
    "\n",
    "location_title = \"############# LOCATION'S MISMACHTS ############## \\n\\n\"\n",
    "gene_title = \"\\n\\n############# GENE_SYMBOL'S MISMACHTS #############\\n\\n\"\n",
    "\n",
    "# Here we introduce a condition, if dicts are empty\n",
    "# write No_error_found\n",
    "# if dicts are. not empty, write all this information\n",
    "\n",
    "if bool(variant2locations) == False:\n",
    "    if bool(gene2gene_API)==False:\n",
    "        with open(completeName,'w') as out:\n",
    "            out.writelines([No_error_found])\n",
    "else:\n",
    "    # Write location\n",
    "    \n",
    "    with open(completeName,'w') as out:\n",
    "            out.writelines([Errors_found, location_title, how_many_location_error, separator, header_location])   \n",
    "    location_error.to_csv(completeName,\n",
    "          header=None, index=None, sep='\\t', mode='a')\n",
    "\n",
    "    \n",
    "    # Now, with Gene_symbol    \n",
    "    \n",
    "    with open(completeName,'a') as out:\n",
    "        out.writelines([gene_title, how_many_GS_error, separator,header_gene_symbol])\n",
    "    gene_error.to_csv(completeName,\n",
    "          header=None, index=None, sep='\\t', mode='a')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# End of stage 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output generated\n",
    "\n",
    "If you are reading this document is because some error(s) or mismatch(s) has been found between the Gene Symbol and/or the location of your variants\n",
    "When these has been compared with API Emsembl.\n",
    "We show you the differences found in this document.\n",
    "\n",
    "############# LOCATION'S MISMACHTS ############## \n",
    "\n",
    "Number of location mismachts found: 1\n",
    "Your location     variant    API location\n",
    "77:43071077\trs1799966\t17:43071077\n",
    "\n",
    "\n",
    "############# GENE_SYMBOL'S MISMACHTS #############\n",
    "\n",
    "Number of Gene Symbol mismachts found: 1\n",
    "Your Gene_symbol,     Variant,       API Gene_symbol\n",
    "BORIS\trs56116432\tABO\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Stage 3\n",
    "****"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [],
   "source": [
    "# In this stage we create the VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The VCF follow a structure explained in https://samtools.github.io/hts-specs/VCFv4.2.pdf\n",
    "# We have created the minimum structure that a VCF document must have \n",
    "# according to documentation (VCFv4.3 format)\n",
    "\n",
    "\n",
    "# An example here\n",
    "\n",
    "#      ##fileformat=VCFv4.3\n",
    "#      #CHROM POS      ID         REF   ALT    QUAL  FILTER   INFO                             \n",
    "#      20     14370    rs6054257  G     A      29    PASS    NS=3;DP=14;AF=0.5;DB;H2\n",
    "\n",
    "\n",
    "\n",
    "# Here we explain how we are going to fill the columns of the  VCF file.\n",
    "\n",
    "### Mandatory columns ###\n",
    "# CHROM taken from df_stage1['CHROM']\n",
    "# POS taken from df_stage1['POS']\n",
    "# ID taken df_stage1['ID']\n",
    "# REF taken from df_stage1['REF']\n",
    "# ALT taken from df_stage1['ALT']\n",
    "# QUAL taken from df_stage1['QUAL']\n",
    "# FILTER taken from df_stage1['FILTER']\n",
    "# INFO filled with (‘.’)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First, let's create a empty document in the same directory\n",
    "\n",
    "# Introce in path the directory where we wish to create the new CVF\n",
    "save_path = '/Users/manolodominguez/Desktop/git-repos/STP_mini_projects/Igor-Manuel/Outputs'\n",
    "\n",
    "# Introduce in name, the name of your new file\n",
    "name_of_file2 = \"The_VCF_file\"\n",
    "\n",
    "# This will  create directory + file name\n",
    "completeName2 = os.path.join(save_path, name_of_file2+\".vcf\")         \n",
    "\n",
    "\n",
    "# Now, let's create the metadata information\n",
    "file_format = '##fileformat=VCFv4.3\\n'\n",
    "today = date.today()\n",
    "d1 = today.strftime(\"%Y%m%d\")\n",
    "file_Date = '##fileDate='+d1+'\\n'\n",
    "\n",
    "# And now the header of the columns\n",
    "header_line = '#CHROM\\tPOS\\tID\\tREF\\tALT\\tQUAL\\tFILTER\\tINFO\\n'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now, we write meta-information line and header line\n",
    "with open(completeName2,'w') as out:\n",
    "    out.writelines([file_format, file_Date,header_line ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>Gene_API</th>\n",
       "      <th>Location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>101232831</td>\n",
       "      <td>NM_016247.3:c.3183C&gt;G</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPG2</td>\n",
       "      <td>3:101232831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>17</td>\n",
       "      <td>43067616</td>\n",
       "      <td>NM_007294.3:c.5066T&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>49/131</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43067616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15</td>\n",
       "      <td>48600225</td>\n",
       "      <td>NM_000138.4:c.356G&gt;A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>46/99</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>FBN1</td>\n",
       "      <td>15:48600225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>14</td>\n",
       "      <td>23426833</td>\n",
       "      <td>NM_000257.3:c.1988G&gt;A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>8/102</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>MYH7</td>\n",
       "      <td>14:23426833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>133256042</td>\n",
       "      <td>rs56116432</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1/200</td>\n",
       "      <td>NO PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>ABO</td>\n",
       "      <td>9:133256042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>17</td>\n",
       "      <td>43071077</td>\n",
       "      <td>rs1799966</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>14/85</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>17:43071077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>-</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>7</td>\n",
       "      <td>117559502</td>\n",
       "      <td>ENST00000003084:c.1431_1433delTTC</td>\n",
       "      <td>TTC</td>\n",
       "      <td>-</td>\n",
       "      <td>99/100</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>CFTR</td>\n",
       "      <td>7:117559502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>NM_002117.5:c.344-46_344-45insT</td>\n",
       "      <td>-</td>\n",
       "      <td>T</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>31271394</td>\n",
       "      <td>rs9281300</td>\n",
       "      <td>-</td>\n",
       "      <td>A</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>6:31271394</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>8</td>\n",
       "      <td>19956018</td>\n",
       "      <td>NM_000237.2:c.953A&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>LPL</td>\n",
       "      <td>8:19956018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>16</td>\n",
       "      <td>2088197</td>\n",
       "      <td>rs1799758</td>\n",
       "      <td>AGTGAG</td>\n",
       "      <td>AG</td>\n",
       "      <td>34/64</td>\n",
       "      <td>PASS</td>\n",
       "      <td>.</td>\n",
       "      <td>TSC2</td>\n",
       "      <td>16:2088197</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM        POS                                 ID     REF ALT    QUAL  \\\n",
       "0      3  101232831              NM_016247.3:c.3183C>G       C   G   34/64   \n",
       "1     17   43067616              NM_007294.3:c.5066T>C       T   C  49/131   \n",
       "2     15   48600225               NM_000138.4:c.356G>A       G   A   46/99   \n",
       "3     14   23426833              NM_000257.3:c.1988G>A       G   A   8/102   \n",
       "4      9  133256042                         rs56116432       C   T   1/200   \n",
       "5     17   43071077                          rs1799966       T   A   14/85   \n",
       "6      7  117559502  ENST00000003084:c.1431_1433delTTC     TTC   -  99/100   \n",
       "7      7  117559502  ENST00000003084:c.1431_1433delTTC     TTC   -  99/100   \n",
       "8      6   31271394    NM_002117.5:c.344-46_344-45insT       -   T   34/64   \n",
       "9      6   31271394                          rs9281300       -   A   34/64   \n",
       "10     8   19956018               NM_000237.2:c.953A>G       A   G   34/64   \n",
       "11    16    2088197                          rs1799758  AGTGAG  AG   34/64   \n",
       "\n",
       "     FILTER INFO Gene_API     Location  \n",
       "0      PASS    .    IMPG2  3:101232831  \n",
       "1      PASS    .    BRCA1  17:43067616  \n",
       "2      PASS    .     FBN1  15:48600225  \n",
       "3   NO PASS    .     MYH7  14:23426833  \n",
       "4   NO PASS    .      ABO  9:133256042  \n",
       "5      PASS    .    BRCA1  17:43071077  \n",
       "6      PASS    .     CFTR  7:117559502  \n",
       "7      PASS    .     CFTR  7:117559502  \n",
       "8      PASS    .    HLA-C   6:31271394  \n",
       "9      PASS    .    HLA-C   6:31271394  \n",
       "10     PASS    .      LPL   8:19956018  \n",
       "11     PASS    .     TSC2   16:2088197  "
      ]
     },
     "execution_count": 351,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Let's modify the df_stage1 to be added to the VCF file\n",
    "df_stage2 = df_stage1\n",
    "df_stage2['INFO'] = '.'\n",
    "\n",
    "df_stage2 = df_stage2[['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER','INFO' ,'Gene_API', 'Location']]\n",
    "df_stage2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_stage2.to_csv(completeName2,\n",
    "          header=None, index=None, sep='\\t', mode='a')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
